---
title: "Biomineralization Toolkit Analysis"
author: "Jill Ashey"
date: "2024-03-08"
output: html_document
---

This script will assess if any biomineralization genes are present in the DEGs. 

A biomineralization gene list was created by FS (primarily from Stylophora pistillata). To see if any of the biomin genes are in this dataset, the gene list was BLASTed against the Pacuta protein sequences. This was done by Zoe Dellaert in this [code](https://github.com/imkristenbrown/Heron-Pdam-gene-expression/blob/master/BioInf/scripts/glmmSeq/analysis/Biomineralization_Toolkit_Analysis.Rmd). Much of what is in this script comes from her code linked above. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(readxl)
```

See [code](https://github.com/JillAshey/JillAshey_Putnam_Lab_Notebook/blob/master/_posts/2024-03-11-BLAST-Pacuta-Hawaii-2022.md) for BLAST information.  

Read in list of biomineralization genes (generated by FS)
```{r}
biomin_genes <- read_excel("../Raw_data/biomineralization_toolkit/Biomineralization_Toolkit_FScucchia.xlsx") %>% 
  dplyr::select(-`blasted protein in Stylophora`)
```

Read in biomineralization blast results for Pacuta (generated by ZD)
```{r}
biomin_blast <- read.delim("../Raw_data/biomineralization_toolkit/Biomineralization_blast_results_tab.txt", header = F) %>%
  dplyr::select(V1, V2) %>%
  distinct()
```

Merge dfs by accession number / gene ID
```{r}
# Merge data frames based on accessionnumber/geneID
merged_data <- biomin_genes %>%
  inner_join(biomin_blast, by = c("accessionnumber/geneID" = "V1")) %>% rename("Pocillopora_acuta_best_hit" = "V2")

#write.csv(merged_data, file = "../output/Biomineralization_Toolkit/biomin_blast_Pacuta_best_hit.csv", row.names = F)
```

How many biomin genes are in our dataset? Look at filtered gene counts first (this is not DEGs, but all of the genes that were expressed in our dataset)
```{r}
counts <- read.csv("../output/DEG/filtered_counts.csv") %>% 
  rename("gene_id" = "X")

# Merge filtered counts and biomin gene information
biomin_filt_genes <- counts %>%
  inner_join(merged_data, by = c("gene_id" = "Pocillopora_acuta_best_hit"))

length(unique(biomin_filt_genes$gene_id))
length(unique(biomin_filt_genes$`accessionnumber/geneID`))
```

158 out of 172 of the biomineralization genes are represented in the filtered gene counts. This corresponds to 84 unique Pacuta genes. 

How many biomin genes are in DEGs?
```{r}
deg <- read.csv("../output/DEG/Unique_DEGs.csv") %>%
  rename("gene_id" = "X")

# Merge DEGs and biomin gene information
biomin_deg_genes <- deg %>%
  inner_join(merged_data, by = c("gene_id" = "Pocillopora_acuta_best_hit"))

length(unique(biomin_deg_genes$gene_id))
length(unique(biomin_deg_genes$`accessionnumber/geneID`))

#write.csv(biomin_deg_genes, file = "../output/Biomineralization_Toolkit/biomin_DEG.csv", row.names = F)
```

7 out of 172 of the biomineralization genes are represented in the DEGs. This corresponds to 5 unique Pacuta genes: 

- Pocillopora_acuta_HIv2___RNAseq.g25214.t1
- Pocillopora_acuta_HIv2___RNAseq.g11609.t1
- Pocillopora_acuta_HIv2___TS.g23498.t1  
- Pocillopora_acuta_HIv2___RNAseq.g7668.t1
- Pocillopora_acuta_HIv2___RNAseq.g30830.t1

These are the 7 biomineralization genes: 

- PFX13778.1
- P33_g8985
- Gene:g38128
- JT016638.1
- P18_g810
- XP_022780303.1
- XP_022805470.1

Where do these 5 unique genes fall in our treatment comparisons? Merge high v control DEGs with biomin gene information
```{r}
## High vs control treatment comparison 
h_v_c <- read.csv("../output/DEG/High_v_Control_DEGs.csv") %>%
  rename("gene_id" = "X")

# Merge DEGs and biomin gene information
biomin_h_v_c <- h_v_c %>%
  inner_join(merged_data, by = c("gene_id" = "Pocillopora_acuta_best_hit"))

length(unique(biomin_h_v_c$gene_id))
unique(biomin_h_v_c$gene_id)

#write.csv(biomin_h_v_c, file = "../output/Biomineralization_Toolkit/biomin_DEG_high_v_control.csv", row.names = F)
```

These Pacuta genes are differentially expressed between high and control treatments and are related to biomineralization: 
- Pocillopora_acuta_HIv2___TS.g23498.t1 - α-Collagen, coadhesion, clone g810 alpha collagen-like protein gene
- Pocillopora_acuta_HIv2___RNAseq.g7668.t1 - uncharacterized protein
- Pocillopora_acuta_HIv2___RNAseq.g30830.t1 - uncharacterized protein

Corals in high treatment are downregulating a collagen related gene compared to control corals (LFC = -3.39). 

Merge mid v control DEGs with biomin gene information
```{r}
## Mid vs control treatment comparison 
m_v_c <- read.csv("../output/DEG/Mid_v_Control_DEGs.csv") %>%
  rename("gene_id" = "X")

# Merge DEGs and biomin gene information
biomin_m_v_c <- m_v_c %>%
  inner_join(merged_data, by = c("gene_id" = "Pocillopora_acuta_best_hit"))

length(unique(biomin_m_v_c$gene_id))
unique(biomin_m_v_c$gene_id)
```

There are no DEGs between mid and control treatments related to biomineralization. 

Merge high v mid DEGs with biomin gene information
```{r}
## High vs control treatment comparison 
h_v_m <- read.csv("../output/DEG/High_v_Mid_DEGs.csv") %>%
  rename("gene_id" = "X")

# Merge DEGs and biomin gene information
biomin_h_v_m <- h_v_m %>%
  inner_join(merged_data, by = c("gene_id" = "Pocillopora_acuta_best_hit"))

length(unique(biomin_h_v_m$gene_id))
unique(biomin_h_v_m$gene_id)

#write.csv(biomin_h_v_m, file = "../output/Biomineralization_Toolkit/biomin_DEG_high_v_mid.csv", row.names = F)
```

These Pacuta genes are differentially expressed between high and mid treatments and are related to biomineralization: 

- Pocillopora_acuta_HIv2___RNAseq.g25214.t1 - Sacsin
- Pocillopora_acuta_HIv2___RNAseq.g11609.t1 - Flagellar associated protein
- Pocillopora_acuta_HIv2___TS.g23498.t1 - α-Collagen, coadhesion, clone g810 alpha collagen-like protein gene
- Pocillopora_acuta_HIv2___RNAseq.g7668.t1 - uncharacterized protein 

Similar to the high v control comparison, corals in high treatment are downregulating a collagen related gene (Pocillopora_acuta_HIv2___TS.g23498.t1) compared to mid corals (LFC = -2.99). Corals in high treatment are also downregulating sacsin (LFC = -1.56) and flagellar-associated protein (LFC = -0.83) compared to mid corals. 

These Pacuta biomin genes were shared between high v control and high v mid DEG comparisons: 

- Pocillopora_acuta_HIv2___TS.g23498.t1 - α-Collagen, coadhesion, clone g810 alpha collagen-like protein gene
- Pocillopora_acuta_HIv2___RNAseq.g7668.t1 - uncharacterized protein

Let's plot these genes!

Read in the unique DEGs file, as that has the normalized counts, and the metadata file
```{r}
degs <- read.csv("../output/DEG/Unique_DEGs.csv")

meta <- read_csv("../Metadata/Pacuta_HI2022_Sequencing_Metadata.csv") %>%
  dplyr::select(Treatment, Species, Lifestage, TubeID) %>%
  mutate(Treatment = ifelse(Treatment == "control", "Control", Treatment)) %>%
  mutate(Treatment = ifelse(Treatment == "medium", "Mid", Treatment)) %>%
  mutate(Treatment = ifelse(Treatment == "high", "High", Treatment)) %>%
  rename(ID = TubeID) %>%
  arrange(ID)
```

Pivot the counts data, join with treatment column, and calculate mean + sd
```{r}
deg_means <- degs %>%
  pivot_longer(cols = -X, names_to = "Treatment", values_to = "Count") %>%
  left_join(meta, by = c("Treatment" = "ID")) %>%
  mutate(log_Count = log(Count)) %>%
  group_by(X, Treatment.y) %>%
  summarize(mean_log_count = mean(log_Count), sd_log_count = sd(log_Count))
```

Make a list of DEGs of interest and filter the df means by that list. Plot counts for DEGs of interest 
```{r}
genes <- list("Pocillopora_acuta_HIv2___TS.g23498.t1", "Pocillopora_acuta_HIv2___RNAseq.g7668.t1")

# Select genes of interest 
filt <- deg_means %>% 
  filter(X %in% genes) %>%
  rename("gene_id" = "X") %>%
  rename("Treatment" = "Treatment.y")  

# Arrange treatment order
filt$Treatment <- factor(filt$Treatment, levels = c("Control", "Mid", "High"))

# Plot 
shared_biomin <- ggplot(filt, aes(x = Treatment, y = mean_log_count, fill = Treatment)) +
  geom_bar(stat = "identity", position = position_dodge(width = 0.9)) +
  # geom_errorbar(aes(ymin = mean_log_count - sd_log_count, ymax = mean_log_count + sd_log_count),
  #               width = 0.2,  # Adjust width as needed
  #               position = position_dodge(width = 0.9)) + # Adjust dodge width as needed
  facet_wrap(~gene_id, scales = "free",) +
  labs(x = "", y = expression(bold(paste("log"[2], " Normalized Counts")))) +
  scale_fill_manual(values = wes_palette("GrandBudapest1", n=3)) +
  #theme_minimal() +
  theme_bw() +
  theme(legend.position = "none") +
  theme(axis.text = element_text(size = 10, color="black"),
        axis.title=element_text(size=12, color="black", face="bold")); shared_biomin
ggsave("../output/Biomineralization_Toolkit/Biomin_shared_DEGs.pdf", shared_biomin)
ggsave("../output/Biomineralization_Toolkit/Biomin_shared_DEGs.png", shared_biomin)

# Plot w/ error bars
shared_biomin_error <- ggplot(filt, aes(x = Treatment, y = mean_log_count, fill = Treatment)) +
  geom_bar(stat = "identity", position = position_dodge(width = 0.9)) +
   geom_errorbar(aes(ymin = mean_log_count - sd_log_count, ymax = mean_log_count + sd_log_count),
                 width = 0.2,  # Adjust width as needed
                 position = position_dodge(width = 0.9)) + # Adjust dodge width as needed
  facet_wrap(~gene_id, scales = "free",) +
  labs(x = "", y = expression(bold(paste("log"[2], " Normalized Counts")))) +
  scale_fill_manual(values = wes_palette("GrandBudapest1", n=3)) +
  #theme_minimal() +
  theme_bw() +
  theme(legend.position = "none") +
  theme(axis.text = element_text(size = 10, color="black"),
        axis.title=element_text(size=12, color="black", face="bold")); shared_biomin_error
ggsave("../output/Biomineralization_Toolkit/Biomin_shared_DEGs_WithErrorBars.pdf", shared_biomin_error)
ggsave("../output/Biomineralization_Toolkit/Biomin_shared_DEGs_WithErrorBars.png", shared_biomin_error)
```


