---
title: "DESeq2"
author: "Jill Ashey"
date: "`r Sys.Date()`"
output: html_document
---

This script will analyze the DEGs from the Pacuta 2022 HI experiment. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(genefilter)
library(DESeq2)
library(pheatmap)
library(lme4)
library(tidyverse)
library(car)
library(gplots)
library(wesanderson)
library(cowplot)
library(gridExtra)
library(grid)
```

Load data. I'm going to use the transcript count matrix first, as that one appears to have most of the info from the gff (whereas the gene count matrix has STRG gene ids)
```{r}
matrix <- read_csv("../output/DEG/Pacuta_transcript_count_matrix.csv")
matrix <- as.data.frame(matrix)
rownames(matrix) <- matrix[,1] #set first column that contains gene names as rownames
matrix <- matrix[,-1] # remove column w/ gene names 
```

Remove extraneous info from sample names 
```{r}
colnames(matrix) <- sub("_R1_001.bam.gtf", "", colnames(matrix))
```

Read in metadata 
```{r}
meta <- read_csv("../Metadata/Pacuta_HI2022_Sequencing_Metadata.csv") %>%
  dplyr::select(Treatment, Species, Lifestage, TubeID) %>%
  mutate(Treatment = ifelse(Treatment == "control", "Control", Treatment)) %>%
  mutate(Treatment = ifelse(Treatment == "medium", "Mid", Treatment)) %>%
  mutate(Treatment = ifelse(Treatment == "high", "High", Treatment)) %>%
  dplyr::rename(ID = TubeID) %>%
  arrange(ID)
```

Set variables as factors 
```{r}
meta$Treatment <- factor(meta$Treatment, levels = c("Control", "Mid", "High"))
```

Data sanity check!
```{r}
meta$ID %in% colnames(matrix) #are all of the sample names (rows) in the metadata df in the gene count matrix? Should be TRUE. 
all(rownames(meta$ID) == colnames(matrix)) #are they the same in the same order? Should be TRUE
```

Filter reads by proportion of samples containing cutoff value 
```{r}
ffun<-filterfun(pOverA(0.25,5))  #set up filtering parameters--LOOK INTO FILTERING PARAMETERS
filt_outrm_poa <- genefilter((matrix), ffun) #apply filter
sum(filt_outrm_poa) #count number of genes left

filt_outrm_poa <- matrix[filt_outrm_poa,] #keep only rows that passed filter

all(rownames(meta$ID) %in% colnames(filt_outrm_poa)) # must come out TRUE

write.csv(filt_outrm_poa, file = "../output/DEG/filtered_counts.csv")
```

Put data and metadata in DESeq2 matrix 
```{r}
data <- DESeqDataSetFromMatrix(countData = filt_outrm_poa, colData = meta, design = ~Treatment)
```

Visualize expression
```{r}
rld <- rlog(data, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
head(assay(rld), 3) #view data
sampleDists <- dist(t(assay(rld))) #calculate distance matix
sampleDistMatrix <- as.matrix(sampleDists) #distance matrix
rownames(sampleDistMatrix) <- colnames(rld) #assign row names
colnames(sampleDistMatrix) <- NULL #assign col names
pheatmap(sampleDistMatrix, #plot matrix of expression similarity
         clustering_distance_rows=sampleDists, #cluster rows
         clustering_distance_cols=sampleDists) #cluster columns

plotPCA(rld, intgroup = c("Treatment")) #plot PCA of samples with all data
```

Run differential gene expression analysis 
```{r}
DEG.int <- DESeq(data) #run differential expression test by group using the wald test 
resultsNames(DEG.int)
```

Look at mid v control contrast
```{r}
m_v_c <- results(DEG.int, contrast = c("Treatment", "Mid", "Control"))
head(m_v_c)
write.csv(m_v_c, file = "../output/DEG/Mid_v_Control_All.csv")
ma_plot <- plotMA(m_v_c); ma_plot
sum(m_v_c$padj <0.05, na.rm = T)
m_v_c.sig <- subset(m_v_c, padj<0.05,) #identify signficant pvalues with 5%FDR
head(m_v_c.sig)
sum(m_v_c.sig$log2FoldChange <0, na.rm=T) # 4 gene is downregulated in mid relative to control
sum(m_v_c.sig$log2FoldChange >0, na.rm=T) # 3 gene are upregulated in mid relative to control
#plotCounts(DEG.int, gene="Pocillopora_acuta_HIv2___RNAseq.g3840.t1", intgroup="Treatment")
m_v_c.sig.list <- data[which(rownames(DEG.int) %in% rownames(m_v_c.sig)),] #subset list of sig transcripts from original count data
m_v_c.rsig <- rlog(m_v_c.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
pca_plot <- plotPCA(m_v_c.rsig, intgroup = c("Treatment")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(m_v_c.rsig) #make an expression object
col.order <- c( "B", "C", "E", "G", # Control samples
                "D", "F", "H", "M", # Mid samples
                "A", "L", "N") # High samples
mat <- mat[,col.order]
df <- as.data.frame(colData(m_v_c.rsig)[,c("Treatment")])
colnames(df) <- "Treatment"
rownames(df) <- c("A", "B", "C", "D", "E", "F", "G", "H", "L", "M", "N")
df <- df[col.order, , drop = FALSE]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=T, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
ggsave("../output/DEG/ma_plot_mid_v_control_DEG.pdf", ma_plot, width = 10, height = 8)
ggsave("../output/DEG/pca_mid_v_control_DEG.pdf", pca_plot, width = 10, height = 8)
ggsave("../output/DEG/heatmap_mid_v_control_DEG.pdf", heatmap, width = 8, height = 10)

# Save info with pvalues and LFC
m_v_c.sig <- as.data.frame(m_v_c.sig)
m_v_c.sig.list <- as.data.frame(counts(m_v_c.sig.list))
m_v_c.sig.list.full <- cbind(m_v_c.sig, m_v_c.sig.list)
write.csv(m_v_c.sig.list.full, file = "../output/DEG/Mid_v_Control_DEGs.csv")
```

Look at high v control contrast
```{r}
h_v_c <- results(DEG.int, contrast = c("Treatment", "High", "Control"))
head(h_v_c)
write.csv(h_v_c, file = "../output/DEG/High_v_Control_All.csv")
ma_plot <- plotMA(h_v_c); ma_plot
sum(h_v_c$padj <0.05, na.rm=T)
h_v_c.sig <- subset(h_v_c, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(h_v_c.sig$log2FoldChange <0, na.rm=T) # 306 genes are downregulated in high relative to control
sum(h_v_c.sig$log2FoldChange >0, na.rm=T) # 366 genes are upregulated in high relative to control
#plotCounts(DEG.int, gene="Pocillopora_acuta_HIv2___RNAseq.g3840.t1", intgroup="Treatment")
h_v_c.sig.list <- data[which(rownames(DEG.int) %in% rownames(h_v_c.sig)),] #subset list of sig transcripts from original count data
h_v_c.rsig <- rlog(h_v_c.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
pca_plot <- plotPCA(h_v_c.rsig, intgroup = c("Treatment")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(h_v_c.rsig) #make an expression object
col.order <- c( "B", "C", "E", "G", # Control samples
                "D", "F", "H", "M", # Mid samples
                "A", "L", "N") # High samples
mat <- mat[,col.order]
df <- as.data.frame(colData(h_v_c.rsig)[,c("Treatment")])
colnames(df) <- "Treatment"
rownames(df) <- c("A", "B", "C", "D", "E", "F", "G", "H", "L", "M", "N")
df <- df[col.order, , drop = FALSE]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=T, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
ggsave("../output/DEG/ma_plot_high_v_control_DEG.pdf", ma_plot, width = 10, height = 8)
ggsave("../output/DEG/pca_high_v_control_DEG.pdf", pca_plot, width = 10, height = 8)
ggsave("../output/DEG/heatmap_high_v_control_DEG.pdf", heatmap, width = 8, height = 10)

# Save info with pvalues and LFC
h_v_c.sig <- as.data.frame(h_v_c.sig)
h_v_c.sig.list <- as.data.frame(counts(h_v_c.sig.list))
h_v_c.sig.list.full <- cbind(h_v_c.sig, h_v_c.sig.list)
write.csv(h_v_c.sig.list.full, file = "../output/DEG/High_v_Control_DEGs.csv")
```

Look at high v mid contrast
```{r}
h_v_m <- results(DEG.int, contrast = c("Treatment", "High", "Mid"))
head(h_v_m)
write.csv(h_v_m, file = "../output/DEG/High_v_Mid_All.csv")
plotMA(h_v_m)
sum(h_v_m$padj <0.05, na.rm=T)
h_v_m.sig <- subset(h_v_m, padj<0.05,) #identify signficant pvalues with 5%FDR
head(h_v_m.sig)
sum(h_v_m.sig$log2FoldChange <0, na.rm=T) # 307 genes are downregulated in high relative to mid
sum(h_v_m.sig$log2FoldChange >0, na.rm=T) # 318 genes are upregulated in high relative to mid
#plotCounts(DEG.int, gene="Pocillopora_acuta_HIv2___RNAseq.g16321.t1", intgroup="Treatment")
h_v_m.sig.list <- data[which(rownames(DEG.int) %in% rownames(h_v_m.sig)),] #subset list of sig transcripts from original count data
write.csv(counts(h_v_m.sig.list), file = "../output/DEG/High_v_Mid_DEGs.csv")
h_v_m.rsig <- rlog(h_v_m.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
pca_plot <- plotPCA(h_v_m.rsig, intgroup = c("Treatment")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(h_v_m.rsig) #make an expression object
col.order <- c( "B", "C", "E", "G", # Control samples
                "D", "F", "H", "M", # Mid samples
                "A", "L", "N") # High samples
mat <- mat[,col.order]
df <- as.data.frame(colData(h_v_m.rsig)[,c("Treatment")])
colnames(df) <- "Treatment"
rownames(df) <- c("A", "B", "C", "D", "E", "F", "G", "H", "L", "M", "N")
df <- df[col.order, , drop = FALSE]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=T, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
ggsave("../output/DEG/ma_plot_high_v_mid_DEG.pdf", ma_plot, width = 10, height = 8)
ggsave("../output/DEG/pca_high_v_mid_DEG.pdf", pca_plot, width = 10, height = 8)
ggsave("../output/DEG/heatmap_high_v_mid_DEG.pdf", heatmap, width = 8, height = 10)

# Save info with pvalues and LFC
h_v_m.sig <- as.data.frame(h_v_m.sig)
h_v_m.sig.list <- as.data.frame(counts(h_v_m.sig.list))
h_v_m.sig.list.full <- cbind(h_v_m.sig, h_v_m.sig.list)
write.csv(h_v_m.sig.list.full, file = "../output/DEG/High_v_Mid_DEGs.csv")
```

The high treatment has some clear expression pattern differences compared to mid and control. 

Look at the unique genes 
```{r}
m_v_c.DEGs <- as.data.frame(row.names(m_v_c.sig.list))
colnames(m_v_c.DEGs) <- "DEGs"
length(t(unique(m_v_c.DEGs)))

h_v_c.DEGs <- as.data.frame(row.names(h_v_c.sig.list))
colnames(h_v_c.DEGs) <- "DEGs"
length(t(unique(h_v_c.DEGs)))

h_v_m.DEGs <- as.data.frame(row.names(h_v_m.sig.list))
colnames(h_v_m.DEGs) <- "DEGs"
length(t(unique(h_v_m.DEGs)))

my_list <- list(m_v_c.DEGs, h_v_c.DEGs, h_v_m.DEGs)
venn(my_list)
# A = Mid v Control; B = High v Control; C = High v Mid

Intersect <- venn(my_list, intersection=TRUE)
isect <- attr(Intersect, "intersection")
str(isect)
```

There is the most overlap between control vs high and mid vs high, suggesting that the high treatment had very different gene expression patterns compared to control and mid. 

Bind the unique genes from all contrasts
```{r}
DEGs.all <- rbind(m_v_c.DEGs, h_v_c.DEGs, h_v_m.DEGs)
DEGs.all <- unique(DEGs.all)
length(unique(DEGs.all$DEGs))

Unique.sig.list <- data[which(rownames(data) %in% DEGs.all$DEGs),] #subset list of sig transcripts from original count data
Unique.rsig <- rlog(Unique.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
write.csv(counts(Unique.sig.list), file="../output/DEG/Unique_DEGs.csv")
```

Visualize the DEGs

PCA with DEGs
```{r}
pca_plot <- plotPCA(Unique.rsig, intgroup = c("Treatment"), returnData = TRUE); pca_plot
percentVar_pca <- round(100*attr(pca_plot, "percentVar")) #plot PCA of samples with all data

pca_plot_degs <- ggplot(pca_plot, aes(PC1, PC2, color = Treatment)) + 
  geom_point(size=15) +
  xlab(paste0("PC1: ",percentVar_pca[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar_pca[2],"% variance")) +
  scale_color_manual(values = wes_palette("GrandBudapest1", n=3)) +
  #ggtitle(label = "M. capitata") +
  theme_bw() + #Set background color
  theme(legend.text = element_text(size=18), 
        #legend.position="none",
        plot.background = element_blank(),
        legend.title = element_text(size=22, face="bold"), 
        #legend.title=element_blank(),
        axis.text = element_text(size=18, colour = "black"), 
        axis.title = element_text(size=20,  face="bold"), 
        axis.title.y = element_text(vjust=-1.5),
        plot.title = element_text(size = 20, face = "italic", hjust = 0.5)) +
  coord_equal(ratio = 2); pca_plot_degs

ggsave("../output/DEG/pca_DEGs.pdf", pca_plot_degs, width = 10, height = 8)
ggsave("../output/DEG/pca_DEGs.png", pca_plot_degs, width = 10, height = 8)
```

Heatmap with DEGs
```{r}
my_colors <- wes_palette("GrandBudapest1", n=3)
red_blue_palette <- colorRampPalette(c("red", "white", "blue"))

mat <- assay(Unique.rsig) #make an expression object
col.order <- c( "B", "C", "E", "G", # Control samples
                "D", "F", "H", "M", # Mid samples
                "A", "L", "N") # High samples
mat <- mat[,col.order]
df <- as.data.frame(colData(Unique.rsig)[,c("Treatment")]) #make dataframe
colnames(df) <- "Treatment"
rownames(df) <- c("A", "B", "C", "D", "E", "F", "G", "H", "L", "M", "N")
df <- df[col.order, , drop = FALSE]

heatmap_degs <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =F,fontsize_row = 2, fontsize = 15, cluster_cols=T, cluster_rows=TRUE, show_colnames =F, annotation_legend = F, annotation_names_col = F, annotation_colors = list(Treatment = c("Control" = my_colors[1], "Mid" = my_colors[2], "High" = my_colors[3]))); heatmap_degs #plot heatmap of all DEG by group

ggsave("../output/DEG/heatmap_DEGs.pdf", heatmap_degs, width = 8, height = 10)
ggsave("../output/DEG/heatmap_DEGs.png", heatmap_degs, width = 8, height = 10)
```

Volcano plots? Include?

HIGH v CONTROL
Add a column to the df that says if the gene is up or down regulated
```{r}
# Add a column to the data frame to specify if they are UP- or DOWN- regulated (log2fc respectively positive or negative)
h_v_c.sig.list.full$diffexpressed <- "NO"

# if log2Foldchange > 1 and adj pvalue < 0.05, set as "UP"
h_v_c.sig.list.full$diffexpressed[h_v_c.sig.list.full$log2FoldChange > 1.0 & h_v_c.sig.list.full$padj < 0.05] <- "UP"

# if log2Foldchange < -1 and adj pvalue < 0.05, set as "UP"
h_v_c.sig.list.full$diffexpressed[h_v_c.sig.list.full$log2FoldChange < -1.0 & h_v_c.sig.list.full$padj < 0.05] <- "DOWN"
```

Make volcano plots for each comparison
```{r}
## Mid v Control
ggplot(h_v_c.sig.list.full, aes(x = log2FoldChange, y = -log10(padj), color = diffexpressed)) + 
  geom_point(size = 2) +
  geom_hline(yintercept = -log10(0.05), col = "black", linetype = 'dashed') +
  geom_vline(xintercept = c(-1.0, 1.0), col = "black", linetype = 'dashed') +
  #geom_text_repel(max.overlaps = Inf, hjust = ifelse(h_v_c_deg$log2FoldChange > 0, 0, 1)) + # To show all labels 
  scale_color_manual(values = c("blue", "grey", "red"), 
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
    scale_fill_manual(values = c("red", "grey", "blue"),  # Match fill colors to diffexpressed
                    labels = c("Downregulated", "Not significant", "Upregulated"),
                    guide = "none") +  # Hide legend for arrow fill colors
  labs(x = expression("log"[2]*"FC"), y = expression("-log"[10]*"p-value")) + 
  theme_classic() +
  theme(axis.title.y = element_text(face = "bold", margin = margin(0,20,0,0), size = rel(1.1), color = 'black'),
        axis.title.x = element_text(hjust = 0.5, face = "bold", margin = margin(20,0,0,0), size = rel(1.1), color = 'black'))
```

HIGH v MID
Add a column to the df that says if the gene is up or down regulated
```{r}
# Add a column to the data frame to specify if they are UP- or DOWN- regulated (log2fc respectively positive or negative)
h_v_m.sig.list.full$diffexpressed <- "NO"

# if log2Foldchange > 1 and adj pvalue < 0.05, set as "UP"
h_v_m.sig.list.full$diffexpressed[h_v_m.sig.list.full$log2FoldChange > 1.0 & h_v_m.sig.list.full$padj < 0.05] <- "UP"

# if log2Foldchange < -1 and adj pvalue < 0.05, set as "UP"
h_v_m.sig.list.full$diffexpressed[h_v_m.sig.list.full$log2FoldChange < -1.0 & h_v_m.sig.list.full$padj < 0.05] <- "DOWN"
```

Make volcano plots for each comparison
```{r}
## Mid v Control
ggplot(h_v_m.sig.list.full, aes(x = log2FoldChange, y = -log10(padj), color = diffexpressed)) + 
  geom_point(size = 2) +
  geom_hline(yintercept = -log10(0.05), col = "black", linetype = 'dashed') +
  geom_vline(xintercept = c(-1.0, 1.0), col = "black", linetype = 'dashed') +
  #geom_text_repel(max.overlaps = Inf, hjust = ifelse(h_v_c_deg$log2FoldChange > 0, 0, 1)) + # To show all labels 
  scale_color_manual(values = c("blue", "grey", "red"), 
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
    scale_fill_manual(values = c("red", "grey", "blue"),  # Match fill colors to diffexpressed
                    labels = c("Downregulated", "Not significant", "Upregulated"),
                    guide = "none") +  # Hide legend for arrow fill colors
  labs(x = expression("log"[2]*"FC"), y = expression("-log"[10]*"p-value")) + 
  theme_classic() +
  theme(axis.title.y = element_text(face = "bold", margin = margin(0,20,0,0), size = rel(1.1), color = 'black'),
        axis.title.x = element_text(hjust = 0.5, face = "bold", margin = margin(20,0,0,0), size = rel(1.1), color = 'black'))
```








Visualize the biomineralization DEGs of interest (identified in [this script](https://github.com/fscucchia/Hawaii2022_pH_Temp_Mcap_Pacu/blob/main/scripts/Biomineralization_Toolkit.Rmd)).
```{r}
# Define the list of gene IDs
gene_ids <- list("Pocillopora_acuta_HIv2___TS.g23498.t1", "Pocillopora_acuta_HIv2___RNAseq.g7668.t1")

# Create a dataframe
genes <- data.frame(gene_id = unlist(gene_ids))

genes <- list("Pocillopora_acuta_HIv2___TS.g23498.t1", "Pocillopora_acuta_HIv2___RNAseq.g7668.t1")
genes <- as.data.frame(genes)

Unique.sig.list.biomin <- data[which(rownames(data) %in% genes$gene_id),] #subset list of sig transcripts from original count data
Unique.rsig.biom <- rlog(Unique.sig.list.biomin, blind=FALSE, fitType = "mean") #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
```

PCA with biomin DEGs 
```{r}
pca_plot <- plotPCA(Unique.rsig.biom, intgroup = c("Treatment"), returnData = TRUE); pca_plot
percentVar_pca <- round(100*attr(pca_plot, "percentVar")) #plot PCA of samples with all data

pca_plot_degs <- ggplot(pca_plot, aes(PC1, PC2, color = Treatment)) + 
  geom_point(size=10) +
  xlab(paste0("PC1: ",percentVar_pca[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar_pca[2],"% variance")) +
  scale_color_manual(values = wes_palette("GrandBudapest1", n=3)) +
  #ggtitle(label = "M. capitata") +
  theme_bw() + #Set background color
  theme(legend.text = element_text(size=13), 
        #legend.position="none",
        plot.background = element_blank(),
        legend.title = element_text(size=18, face="bold"), 
        #legend.title=element_blank(),
        axis.text = element_text(size=13), 
        axis.title = element_text(size=15,  face="bold"), 
        axis.title.y = element_text(vjust=-1.5),
        plot.title = element_text(size = 20, face = "italic", hjust = 0.5)) +
  coord_equal(ratio = 2); pca_plot_degs
```

Heatmap with biomin DEGs
```{r}
my_colors <- wes_palette("GrandBudapest1", n=3)

mat <- assay(Unique.rsig.biom) #make an expression object
col.order <- c( "B", "C", "E", "G", # Control samples
                "D", "F", "H", "M", # Mid samples
                "A", "L", "N") # High samples
mat <- mat[,col.order]
df <- as.data.frame(colData(Unique.rsig.biom)[,c("Treatment")]) #make dataframe
colnames(df) <- "Treatment"
rownames(df) <- c("A", "B", "C", "D", "E", "F", "G", "H", "L", "M", "N")
df <- df[col.order, , drop = FALSE]

heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =T,fontsize_row = 8, cluster_cols=T, cluster_rows=TRUE, show_colnames =F, annotation_legend = T, annotation_names_col = F, annotation_colors = list(Treatment = c("Control" = my_colors[1], "Mid" = my_colors[2], "High" = my_colors[3]))); heatmap #plot heatmap of all DEG by group

ggsave("../output/Biomineralization_Toolkit/heatmap_biomin_DEGs.pdf", heatmap, width = 8, height = 10)
ggsave("../output/Biomineralization_Toolkit/heatmap_biomin_DEGs.png", heatmap, width = 8, height = 10)
```

