---
title: "Enrichment analysis"
author: "Jill Ashey"
date: "`r Sys.Date()`"
output: html_document
---

This script will perform functional enrichment analysis with GO and KEGG terms using the [identified DEGs](https://github.com/fscucchia/Hawaii2022_pH_Temp_Mcap_Pacu/blob/main/scripts/DESeq2.Rmd) from the Pacuta 2022 HI experiment. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load libraries
library("DESeq2")
library("tidyverse")
library("dplyr")
library("pheatmap")
library("RColorBrewer")
library("genefilter")
library("ggplot2")
library("gplots")
library("limma")
library("spdep") 
library("adegenet") 
library("goseq")
library("gridExtra")
library("clusterProfiler")
library("DataCombine")
library("VennDiagram")
library(rtracklayer)
#library(rrvgo)
library(GO.db)
library(tidyr)
library(forcats)
library(wesanderson)
library(GSEABase)
library(directlabels)
library(ggrepel)

#devtools::install_github("krlmlr/ulimit")
#library(ulimit)

#ulimit::memory_limit(7970592)
```

Read in all expressed genes (poverA = 0.75, 5)
```{r}
filt_counts <- read.csv("../output/DEG/filtered_counts.csv")
colnames(filt_counts)[1] <- "gene_id"

str(filt_counts)
```

Read in DEGs
```{r}
degs <- read.csv("../output/DEG/Unique_DEGs.csv")
colnames(degs)[1] <- "gene_id"

str(degs)
```

Calculate read lengths from Pacuta V2 [cds file](http://cyanophora.rutgers.edu/Pocillopora_acuta/). I did this on my local computer using the code below. 
```{bash}
cd /Users/jillashey/Desktop/GFFs/pacuta/V2

awk 'BEGIN{FS="[> ]"} /^>/{val=$2;next}  {print val,length($0)}' Pocillopora_acuta_HIv2.genes.cds.fna > gene_length.txt
```

Read in gene lengths 
```{r}
length <- read.csv("../output/DEG/gene_length.txt", header = F)

length <- length %>%
  separate(V1, into = c("gene_id", "length"), sep = " ", remove = FALSE) %>%
  dplyr::select(-V1)

str(length)
```

Merge length data with filtered count data
```{r}
length_merge <- merge(length, filt_counts, by = "gene_id")

str(length_merge)
```

GOseq requires a vector of all genes, all differentially expressed genes, and gene lengths. Make vectors 
```{r}
# Make DEG vector
DEG <- length_merge[length_merge$gene_id %in% degs$gene_id, ]
DEG_names <- as.vector(DEG$gene_id)
gene_vector <- as.integer(length_merge$gene_id%in%DEG_names)
names(gene_vector) <- length_merge$gene_id

# Make ID vector
ID_vector <- as.vector(length_merge$gene_id)

# Make length vector
length_vector <- as.numeric(length_merge$length)
```

Calculate probability weighting function 
```{r}
DEG.pwf <- nullp(gene_vector, ID_vector, bias.data = length_vector)

str(DEG.pwf)
```

Code above giving me this warning: Warning: initial point very close to some inequality constraintsWarning: collapsing to unique 'x' values

Read in GO annotation information
```{r}
annot <- read.delim("../Metadata/Pocillopora_acuta_HIv2.genes.EggNog_results.txt", header = T)
```

This annot file has GO, Kegg, eggNog and Pfam annotation information. I'm going to start analyzing the GO terms first.

Select GO columns and split so that there is only one GO term per row
```{r}
annot_GO <- annot %>%
  dplyr::select(X.query, GOs) %>%
  mutate_all(~if_else(. == "-", NA_character_, .)) %>%
  separate_rows(GOs, sep = ",") %>%
  rename(gene_id = X.query,
         GO.ID = GOs)

# Remove leading and trailing whitespaces
annot_GO$GO.ID <- trimws(annot_GO$GO.ID)

# Replace NAs with unknown
annot_GO$GO.ID <- replace_na(annot_GO$GO.ID, "unknown")

# Set gene and go ids as factors 
annot_GO$gene_id <- as.factor(annot_GO$gene_id)
annot_GO$GO.ID <- as.factor(annot_GO$GO.ID)

# Select only unique rows
annot_GO <- unique(annot_GO)

# Look at unique gene and go ids 
length(unique(annot_GO$gene_id))
length(unique(annot_GO$GO.ID))

# Remove rows with unknowns
# annot_GO <- annot_GO %>%
#   filter(GO.ID != "unknown")
```

Remove the annotation information for genes not present in our data 
```{r}
filt_annot_GO <- subset(annot_GO, gene_id %in% filt_counts$gene_id)

str(filt_annot_GO)

# Convert filt_annot_GO from tibble to df 
filt_annot_GO <-as.data.frame(filt_annot_GO)
```


Perform GOSeq
```{r}
GO.wall<-goseq::goseq(DEG.pwf, ID_vector, gene2cat=filt_annot_GO, method="Wallenius", use_genes_without_cat=T)

GO <- GO.wall[order(GO.wall$over_represented_pvalue),]
write.csv(GO, file = "../output/functional_enrichment/GOSeq_GO_all.csv")
```

Worked! 

Filter by significantly enriched GO terms (p < 0.05)
```{r}
GO_05 <- GO %>%
  dplyr::filter(over_represented_pvalue<0.05) %>%
        dplyr::arrange(., ontology, over_represented_pvalue)

write.csv(GO_05, file = "../output/functional_enrichment/GOSeq_GO_05.csv")
```

Plot all ontologies (BP, CC, MF) and order by p-value 
```{r}
GO_05_plot <- GO_05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_text(aes(label = over_represented_pvalue), hjust = -1, vjust = 0, size = 2) +
  geom_point(size=1, aes(colour = ontology)) +
  coord_flip() +
  ylim(0,305) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); GO_05_plot #Set the plot background #set title attributes

ggsave("../output/functional_enrichment/GOSeq_GO_05.pdf", GO_05_plot, width = 20, height = 10)
ggsave("../output/functional_enrichment/GOSeq_GO_05.png", GO_05_plot, width = 20, height = 10)
```

Plot ontologies (BP, CC, MF) separately and order by p-value 
```{r}
## BP 
GO_05_plot_BP <- GO_05 %>% 
  drop_na(ontology) %>% 
  filter(ontology == "BP") %>%
  mutate(term = fct_reorder(term, numDEInCat)) %>%
  #mutate(term = fct_reorder(term)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_text(aes(label = over_represented_pvalue), hjust = -1, vjust = 0, size = 2) +
  geom_point(size=1) +
  coord_flip() +
  ylim(0,305) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); GO_05_plot_BP #Set the plot background #set title attributes

ggsave("../output/functional_enrichment/GOSeq_GO_05_BP.pdf", GO_05_plot_BP, width = 30, height = 10)
ggsave("../output/functional_enrichment/GOSeq_GO_05_BP.png", GO_05_plot_BP, width = 30, height = 10)

## CC
GO_05_plot_CC <- GO_05 %>% 
  drop_na(ontology) %>% 
  filter(ontology == "CC") %>%
  mutate(term = fct_reorder(term, numDEInCat)) %>%
  #mutate(term = fct_reorder(term)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_text(aes(label = over_represented_pvalue), hjust = -1, vjust = 0, size = 2) +
  geom_point(size=1) +
  coord_flip() +
  ylim(0,305) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); GO_05_plot_CC #Set the plot background #set title attributes

ggsave("../output/functional_enrichment/GOSeq_GO_05_CC.pdf", GO_05_plot_CC, width = 30, height = 10)
ggsave("../output/functional_enrichment/GOSeq_GO_05_CC.png", GO_05_plot_CC, width = 30, height = 10)

## MF
GO_05_plot_MF <- GO_05 %>% 
  drop_na(ontology) %>% 
  filter(ontology == "MF") %>%
  mutate(term = fct_reorder(term, numDEInCat)) %>%
  #mutate(term = fct_reorder(term)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_text(aes(label = over_represented_pvalue), hjust = -1, vjust = 0, size = 2) +
  geom_point(size=1) +
  coord_flip() +
  ylim(0,305) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); GO_05_plot_MF #Set the plot background #set title attributes

ggsave("../output/functional_enrichment/GOSeq_GO_05_MF.pdf", GO_05_plot_MF, width = 30, height = 10)
ggsave("../output/functional_enrichment/GOSeq_GO_05_MF.png", GO_05_plot_MF, width = 30, height = 10)
```

Merge significant GO terms and annot list together 
```{r}
GO_genes <- annot_GO %>%
  inner_join(GO_05, by = c("GO.ID" = "category"))

length(unique(GO_genes$gene_id)) # 3786 genes with over-represented GO terms 
```

Read in DEGs and merge that data with significant GO terms 
```{r}
deg <- read.csv("../output/DEG/Unique_DEGs.csv") %>%
  rename("gene_id" = "X")

# Merge DEGs and biomin gene information
GO_deg <- GO_genes %>%
  inner_join(deg, by = "gene_id")

length(unique(GO_deg$gene_id)) # 204 DEGs with over-represented GO terms 

write.csv(GO_deg, file = "../output/functional_enrichment/DEG_GO.csv")
```

How does this relate to our DEG comparisons? Merge high v control DEGs with GO information
```{r}
## High vs control treatment comparison 
h_v_c <- read.csv("../output/DEG/High_v_Control_DEGs.csv") %>%
  rename("gene_id" = "X")

# Merge DEGs and GO information and remove extra columns
GO_h_v_c <- h_v_c %>%
  inner_join(GO_deg, by = "gene_id") %>%
  dplyr::select(-c("A.y", "B.y", "C.y", "D.y", "E.y", "F.y", "G.y", "H.y", "L.y", "M.y", "N.y"))

length(unique(GO_h_v_c$gene_id))

write.csv(GO_h_v_c, file = "../output/functional_enrichment/High_v_Control_GO.csv")
```

Merge Mid v control DEGs with GO information
```{r}
## Mid vs control treatment comparison 
m_v_c <- read.csv("../output/DEG/Mid_v_Control_DEGs.csv") %>%
  rename("gene_id" = "X")

# Merge DEGs and GO information and remove extra columns
GO_m_v_c <- m_v_c %>%
  inner_join(GO_deg, by = "gene_id") %>%
  dplyr::select(-c("A.y", "B.y", "C.y", "D.y", "E.y", "F.y", "G.y", "H.y", "L.y", "M.y", "N.y"))

length(unique(GO_m_v_c$gene_id))

write.csv(GO_m_v_c, file = "../output/functional_enrichment/Mid_v_Control_GO.csv")
```

Merge high v mid DEGs with GO information
```{r}
## High vs mid treatment comparison 
h_v_m <- read.csv("../output/DEG/High_v_Mid_DEGs.csv") %>%
  rename("gene_id" = "X")

# Merge DEGs and GO information and remove extra columns
GO_h_v_m <- h_v_m %>%
  inner_join(GO_deg, by = "gene_id") %>%
  dplyr::select(-c("A.y", "B.y", "C.y", "D.y", "E.y", "F.y", "G.y", "H.y", "L.y", "M.y", "N.y"))

length(unique(GO_h_v_m$gene_id))

write.csv(GO_h_v_m, file = "../output/functional_enrichment/High_v_Mid_GO.csv")
```

Let's visualize these data. First, High v Control. Positive LFC means that the genes are upregulated in the high treatment. 
```{r}
GO_BP_High_v_Control <- GO_h_v_c %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "BP") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_BP_High_v_Control
ggsave("../output/functional_enrichment/GO_BP_High_v_Control_top50_pvalue.pdf", GO_BP_High_v_Control, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_BP_High_v_Control_top50_pvalue.png", GO_BP_High_v_Control, width = 30, height = 10)

GO_CC_High_v_Control <- GO_h_v_c %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "CC") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_CC_High_v_Control
ggsave("../output/functional_enrichment/GO_CC_High_v_Control_top50_pvalue.pdf", GO_CC_High_v_Control, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_CC_High_v_Control_top50_pvalue.png", GO_CC_High_v_Control, width = 30, height = 10)

GO_MF_High_v_Control <- GO_h_v_c %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "MF") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=log2FoldChange, 
               size=pvalue)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_MF_High_v_Control
ggsave("../output/functional_enrichment/GO_MF_High_v_Control_top50_pvalue.pdf", GO_MF_High_v_Control, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_MF_High_v_Control_top50_pvalue.png", GO_MF_High_v_Control, width = 30, height = 10)
```

Mid v Control. Positive LFC means that the genes are upregulated in the mid treatment. Mid v Control only shares 1 gene with GO terms and all GO terms are BP. 
```{r}
GO_BP_Mid_v_Control <- GO_m_v_c %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "BP") %>%
    #dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_BP_Mid_v_Control
ggsave("../output/functional_enrichment/GO_BP_Mid_v_Control_top50_pvalue.pdf", GO_BP_Mid_v_Control, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_BP_Mid_v_Control_top50_pvalue.png", GO_BP_Mid_v_Control, width = 30, height = 10)
```

High v Mid Positive LFC means that the genes are upregulated in the high treatment. 
```{r}
GO_BP_High_v_Mid <- GO_h_v_m %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "BP") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_BP_High_v_Mid
ggsave("../output/functional_enrichment/GO_BP_High_v_Mid_top50_pvalue.pdf", GO_BP_High_v_Mid, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_BP_High_v_Mid_top50_pvalue.png", GO_BP_High_v_Mid, width = 30, height = 10)

GO_CC_High_v_Mid <- GO_h_v_m %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "CC") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_CC_High_v_Mid
ggsave("../output/functional_enrichment/GO_CC_High_v_Mid_top50_pvalue.pdf", GO_CC_High_v_Mid, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_CC_High_v_Mid_top50_pvalue.png", GO_CC_High_v_Mid, width = 30, height = 10)

GO_MF_High_v_Mid <- GO_h_v_m %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "MF") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_MF_High_v_Mid
ggsave("../output/functional_enrichment/GO_MF_High_v_Mid_top50_pvalue.pdf", GO_MF_High_v_Mid, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_MF_High_v_Mid_top50_pvalue.png", GO_MF_High_v_Mid, width = 30, height = 10)
```

Join the GO High v Control df and GO High v Mid df by gene
```{r}
# Remove extra info
# GO_h_v_c <- GO_h_v_c %>%
#     dplyr::select(-c("baseMean", "lfcSE", "stat", "A.x", "B.x", "C.x", "D.x", "E.x", "F.x", "G.x", "H.x", "L.x", "M.x", "N.x", "under_represented_pvalue")) %>%
#     rename_with(~ paste0("h_v_c_", .), everything())
# 
# 
# GO_h_v_m <- GO_h_v_m %>%
#     dplyr::select(-c("baseMean", "lfcSE", "stat", "A.x", "B.x", "C.x", "D.x", "E.x", "F.x", "G.x", "H.x", "L.x", "M.x", "N.x", "under_represented_pvalue")) %>%
#     rename_with(~ paste0("h_v_m_", .), everything())
# 
# 
# test <- GO_h_v_c %>%
#   inner_join(GO_h_v_m, by = c("h_v_c_gene_id" = "h_v_m_gene_id"))
```

Getting this warming: Warning: Detected an unexpected many-to-many relationship between `x` and `y`.
. How can a gene have different enriched GO terms depending on the treatment?? 

Manual comparisons for GO High v Control df and GO High v Mid df
```{r}
GO_h_v_c <- GO_h_v_c %>%
  na.omit() %>%
  dplyr::arrange(log2FoldChange) %>%
  print()

GO_h_v_m <- GO_h_v_m %>%
  na.omit() %>%
  dplyr::arrange(log2FoldChange) %>%
  print()
```

Since there are so issues when joining the files, I'm going to look at the gene ids shared between these two treatment comparisons 
```{r}
my_list <- list(unique(GO_h_v_c$gene_id), unique(GO_h_v_m$gene_id))
venn(my_list)
Intersect <- venn(my_list, intersection=TRUE)
isect <- attr(Intersect, "intersection")
str(isect)
isect$`A:B`
```

Filter the GO dfs so that we only have the shared genes. 

```{r}
GO_h_v_c_filt <- GO_h_v_c %>%
  dplyr::filter(gene_id %in% isect$`A:B`)

GO_h_v_m_filt <- GO_h_v_m %>%
  dplyr::filter(gene_id %in% isect$`A:B`)

# Check to see if gene_id columns are identical
identical(GO_h_v_c_filt$gene_id, GO_h_v_m_filt$gene_id) # they are not 
```

*NOTE*: the genes below are only those that are shared between h v c and h v m comparisons 

The top 5 genes that are most DOWNregulated in the high treatment compared to the control treatment are:

- Pocillopora_acuta_HIv2___RNAseq.g8481.t1 (LFC -8.8221765)
- Pocillopora_acuta_HIv2___RNAseq.g12512.t1 (LFC -8.446681)
- Pocillopora_acuta_HIv2___RNAseq.22237_t (LFC -7.667258)
- Pocillopora_acuta_HIv2___RNAseq.g24121.t1 (LFC -5.6686604)
- Pocillopora_acuta_HIv2___RNAseq.g13974.t1 (LFC -4.1565947)

The top 5 genes that are most DOWNregulated in the high treatment compared to the mid treatment are:

- Pocillopora_acuta_HIv2___RNAseq.g8481.t1 (LFC -8.396540)
- Pocillopora_acuta_HIv2___RNAseq.g12512.t1 (LFC -7.753341)
- Pocillopora_acuta_HIv2___RNAseq.22237_t (LFC -7.162973)
- Pocillopora_acuta_HIv2___RNAseq.g24121.t1 (LFC -5.554880)
- Pocillopora_acuta_HIv2___RNAseq.g13974.t1 (LFC -4.392526)

The top 5 genes that are most UPregulated in the high treatment compared to the control treatment are:

- Pocillopora_acuta_HIv2___RNAseq.g22884.t1 (LFC 4.549559)
- Pocillopora_acuta_HIv2___TS.g23786.t1 (LFC 3.653126)
- Pocillopora_acuta_HIv2___RNAseq.3335_t (LFC 3.001841)
- Pocillopora_acuta_HIv2___RNAseq.g19477.t1 (LFC 2.848650)
- Pocillopora_acuta_HIv2___RNAseq.g19484.t2 (LFC 2.744931)

The top 5 genes that are most UPregulated in the high treatment compared to the mid treatment are:

- Pocillopora_acuta_HIv2___RNAseq.g22884.t1 (LFC 4.709843)
- Pocillopora_acuta_HIv2___RNAseq.3335_t (LFC 4.278247)
- Pocillopora_acuta_HIv2___TS.g23786.t1 (LFC 3.599031)
- Pocillopora_acuta_HIv2___RNAseq.g19477.t1 (LFC 3.373083)
- Pocillopora_acuta_HIv2___TS.g26760.t1 (LFC 2.893989)

Compared to mid and control corals, high corals strongly downregulated `Pocillopora_acuta_HIv2___RNAseq.g8481.t1`, `Pocillopora_acuta_HIv2___RNAseq.g12512.t1`, `Pocillopora_acuta_HIv2___RNAseq.22237_t`, `Pocillopora_acuta_HIv2___RNAseq.g24121.t1`, and `Pocillopora_acuta_HIv2___RNAseq.g13974.t1`. 

Compared to mid and control corals, high corals strongly upregulated `Pocillopora_acuta_HIv2___RNAseq.g22884.t1`, `Pocillopora_acuta_HIv2___TS.g23786.t1`, `Pocillopora_acuta_HIv2___RNAseq.g19477.t1`, and `Pocillopora_acuta_HIv2___RNAseq.3335_t`.

Given these genes have the greatest LFC in the comparisons, I think we should investigate them further. 

Plot the DEGs of interest with the GO information. First, read in DEG and GO information.
```{r}
deg_go <- read.csv("../output/functional_enrichment/DEG_GO.csv")
```

Filter by downregulated genes of interest 
```{r}
down <- c("Pocillopora_acuta_HIv2___RNAseq.g8481.t1", "Pocillopora_acuta_HIv2___RNAseq.g12512.t1", "Pocillopora_acuta_HIv2___RNAseq.22237_t", "Pocillopora_acuta_HIv2___RNAseq.g24121.t1", "Pocillopora_acuta_HIv2___RNAseq.g13974.t1")

# Select genes of interest 
down_filt <- deg_go %>% 
  filter(gene_id %in% down) %>%
  na.omit()
```

Plot GO term information for downregulated genes of interest 
```{r}
down_deg_go <- down_filt %>% 
    mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    na.omit() %>%
    #dplyr::filter(ontology == "BP") %>%
    #dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    #top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               #colour=pvalue, 
               color=over_represented_pvalue)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", size="pvalue"); down_deg_go
```

Filter by upregulated genes of interest 
```{r}
up <- c("Pocillopora_acuta_HIv2___RNAseq.g22884.t1", "Pocillopora_acuta_HIv2___TS.g23786.t1", "Pocillopora_acuta_HIv2___RNAseq.g19477.t1", "Pocillopora_acuta_HIv2___RNAseq.3335_t")

# Select genes of interest 
up_filt <- deg_go %>% 
  filter(gene_id %in% up) %>%
  na.omit()
```

Plot GO term information for downregulated genes of interest 
```{r}
up_deg_go <- up_filt %>% 
    mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    na.omit() %>%
    #dplyr::filter(ontology == "BP") %>%
    #dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    #top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               #colour=pvalue, 
               color=over_represented_pvalue)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", size="pvalue"); up_deg_go
```

Add slim terms to dataset 
```{r}
GOslim <- read_csv("~/Desktop/PutnamLab/Repositories/SedimentStress/SedimentStress/Output/GOSeq/GO-GOslim.csv")
```

Merge GO slim info and log p-value
```{r}
down_filt_slim <- down_filt %>%
  inner_join(GOslim, by = c("GO.ID" = "GO_ID"))

# Log the overrepresented pvalue 
down_filt_slim <- down_filt_slim %>%
  mutate(log_over_represented_pvalue = -log10(over_represented_pvalue))

# Sort the data by GOSlim_bin and log_over_represented_pvalue
down_filt_slim <- down_filt_slim %>%
  arrange(GOSlim_bin, log_over_represented_pvalue)

# Save csv with slim info for these genes 
write.csv(down_filt_slim, file = "../output/functional_enrichment/Down_filt_genes_GOslim.csv")
```

Plot downregulated GO slim information. 
```{r}
# Set the factor levels for term.x based on the sorted order
down_filt_slim$term.x <- factor(down_filt_slim$term.x, levels = unique(down_filt_slim$term.x))

# Set the order of the facet grid levels
down_filt_slim$ontology <- factor(down_filt_slim$ontology, levels = c("BP", "MF", "CC"))

# Create a new factor for GOSlim_bin ordered by ontology
down_filt_slim <- down_filt_slim %>%
  mutate(GOSlim_bin = factor(GOSlim_bin, levels = unique(GOSlim_bin[order(ontology)])))

# Plot
down_filt_slim_plot <- down_filt_slim %>%
      ggplot(aes(x=log_over_represented_pvalue, 
               y=term.x,
               colour=ontology)) +
        geom_point(size = 20) +
    scale_color_manual(values = c("cornflowerblue", "mediumvioletred", "seagreen")) +
        #expand_limits(x=0) +
        #labs(x="Number of DEGs", y="GO term", color="pvalue") %>%
        facet_grid(GOSlim_bin ~ ., scales = "free", space='free', labeller = label_wrap_gen(width = 55, multi_line = TRUE)) +
      #xlab("Number of DEGs") +
      #ylab("") +
      labs(x="-log10(over-represented p-value)", y="") +
      theme_bw() + 
   theme(
    strip.text.y = element_text(size = 28, colour = "black", angle = 0, face = "bold"),
    axis.text = element_text(size = 32, colour = "black"),
    axis.title.x = element_text(size = 35, colour = "black"),
    legend.text = element_text(size = 24),  # Increase legend text size
    legend.title = element_text(size = 26), # Increase legend title size
    legend.key.size = unit(1.5, "lines"),   # Increase size of legend key
    panel.border = element_rect(colour = "black", size = 1.5),  # Darken and thicken the panel border
    strip.background = element_rect(colour = "black", size = 1.5) # Darken and thicken the strip background border
  )

ggsave("../output/functional_enrichment/Downregulated_shared_genes_GOSlim.pdf", down_filt_slim_plot, width = 35, height = 25)
ggsave("../output/functional_enrichment/Downregulated_shared_genes_GOSlim.png", down_filt_slim_plot, width = 35, height = 25)
```

Merge GO slim info and log p-value
```{r}
up_filt_slim <- up_filt %>%
  inner_join(GOslim, by = c("GO.ID" = "GO_ID"))

# Log the overrepresented pvalue 
up_filt_slim <- up_filt_slim %>%
  mutate(log_over_represented_pvalue = -log10(over_represented_pvalue))

# Sort the data by GOSlim_bin and log_over_represented_pvalue
up_filt_slim <- up_filt_slim %>%
  arrange(GOSlim_bin, log_over_represented_pvalue)

# Save csv with slim info for these genes 
#write.csv(up_filt_slim, file = "../output/functional_enrichment/Up_filt_genes_GOslim.csv")
```

Plot
```{r}
# Set the factor levels for term.x based on the sorted order
up_filt_slim$term.x <- factor(up_filt_slim$term.x, levels = unique(up_filt_slim$term.x))

# Set the order of the facet grid levels
up_filt_slim$ontology <- factor(up_filt_slim$ontology, levels = c("BP", "MF", "CC"))

# Create a new factor for GOSlim_bin ordered by ontology
up_filt_slim <- up_filt_slim %>%
  mutate(GOSlim_bin = factor(GOSlim_bin, levels = unique(GOSlim_bin[order(ontology)])))

# Plot
up_filt_slim_plot <- up_filt_slim %>%
      ggplot(aes(x=log_over_represented_pvalue, 
               y=term.x,
               colour=ontology)) +
        geom_point(size = 20) +
      scale_color_manual(values = c("cornflowerblue", "mediumvioletred", "seagreen")) +
        #expand_limits(x=0) +
        facet_grid(GOSlim_bin ~ ., scales = "free", space='free', labeller = label_wrap_gen(width = 55, multi_line = TRUE)) +
      #xlab("Number of DEGs") +
      #ylab("") +
      labs(x="-log10(over-represented p-value)", y="") +
      theme_bw() + 
   theme(
    strip.text.y = element_text(size = 28, colour = "black", angle = 0, face = "bold"),
    axis.text = element_text(size = 32, colour = "black"),
    axis.title.x = element_text(size = 35, colour = "black"),
    legend.text = element_text(size = 24),  # Increase legend text size
    legend.title = element_text(size = 26), # Increase legend title size
    legend.key.size = unit(1.5, "lines"),   # Increase size of legend key
    panel.border = element_rect(colour = "black", size = 1.5),  # Darken and thicken the panel border
    strip.background = element_rect(colour = "black", size = 1.5) # Darken and thicken the strip background border
  )

ggsave("../output/functional_enrichment/Upregulated_shared_genes_GOSlim.pdf", up_filt_slim_plot, width = 35, height = 25)
ggsave("../output/functional_enrichment/Upregulated_shared_genes_GOSlim.png", up_filt_slim_plot, width = 35, height = 25)
```





