---
title: "Enrichment analysis"
author: "Jill Ashey"
date: "`r Sys.Date()`"
output: html_document
---

This script will perform functional enrichment analysis with GO and KEGG terms using the [identified DEGs](https://github.com/fscucchia/Hawaii2022_pH_Temp_Mcap_Pacu/blob/main/scripts/DESeq2.Rmd) from the Pacuta 2022 HI experiment. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load libraries
library("DESeq2")
library("tidyverse")
library("dplyr")
library("pheatmap")
library("RColorBrewer")
library("genefilter")
library("ggplot2")
library("gplots")
library("limma")
library("spdep") 
library("adegenet") 
library("goseq")
library("gridExtra")
library("clusterProfiler")
library("DataCombine")
library("VennDiagram")
library(rtracklayer)
#library(rrvgo)
library(GO.db)
library(tidyr)
library(forcats)
library(wesanderson)

#devtools::install_github("krlmlr/ulimit")
#library(ulimit)

#ulimit::memory_limit(7970592)
```

Read in all expressed genes (poverA = 0.75, 5)
```{r}
filt_counts <- read.csv("../output/DEG/filtered_counts.csv")
colnames(filt_counts)[1] <- "gene_id"

str(filt_counts)
```

Read in DEGs
```{r}
degs <- read.csv("../output/DEG/Unique_DEGs.csv")
colnames(degs)[1] <- "gene_id"

str(degs)
```

Calculate read lengths from Pacuta V2 [cds file](http://cyanophora.rutgers.edu/Pocillopora_acuta/). I did this on my local computer using the code below. 
```{bash}
cd /Users/jillashey/Desktop/GFFs/pacuta/V2

awk 'BEGIN{FS="[> ]"} /^>/{val=$2;next}  {print val,length($0)}' Pocillopora_acuta_HIv2.genes.cds.fna > gene_length.txt
```

Read in gene lengths 
```{r}
length <- read.csv("../output/DEG/gene_length.txt", header = F)

length <- length %>%
  separate(V1, into = c("gene_id", "length"), sep = " ", remove = FALSE) %>%
  dplyr::select(-V1)

str(length)
```

Merge length data with filtered count data
```{r}
length_merge <- merge(length, filt_counts, by = "gene_id")

str(length_merge)
```

GOseq requires a vector of all genes, all differentially expressed genes, and gene lengths. Make vectors 
```{r}
# Make DEG vector
DEG <- length_merge[length_merge$gene_id %in% degs$gene_id, ]
DEG_names <- as.vector(DEG$gene_id)
gene_vector <- as.integer(length_merge$gene_id%in%DEG_names)
names(gene_vector) <- length_merge$gene_id

# Make ID vector
ID_vector <- as.vector(length_merge$gene_id)

# Make length vector
length_vector <- as.numeric(length_merge$length)
```

Calculate probability weighting function 
```{r}
DEG.pwf <- nullp(gene_vector, ID_vector, bias.data = length_vector)

str(DEG.pwf)
```

Code above giving me this warning: Warning: initial point very close to some inequality constraintsWarning: collapsing to unique 'x' values

Read in GO annotation information
```{r}
annot <- read.delim("../Metadata/Pocillopora_acuta_HIv2.genes.EggNog_results.txt", header = T)
```

This annot file has GO, Kegg, eggNog and Pfam annotation information. I'm going to start analyzing the GO terms first.

Select GO columns and split so that there is only one GO term per row
```{r}
annot_GO <- annot %>%
  dplyr::select(X.query, GOs) %>%
  mutate_all(~if_else(. == "-", NA_character_, .)) %>%
  separate_rows(GOs, sep = ",") %>%
  rename(gene_id = X.query,
         GO.ID = GOs)

# Remove leading and trailing whitespaces
annot_GO$GO.ID <- trimws(annot_GO$GO.ID)

# Replace NAs with unknown
annot_GO$GO.ID <- replace_na(annot_GO$GO.ID, "unknown")

# Set gene and go ids as factors 
annot_GO$gene_id <- as.factor(annot_GO$gene_id)
annot_GO$GO.ID <- as.factor(annot_GO$GO.ID)

# Select only unique rows
annot_GO <- unique(annot_GO)

# Look at unique gene and go ids 
length(unique(annot_GO$gene_id))
length(unique(annot_GO$GO.ID))

# Remove rows with unknowns
# annot_GO <- annot_GO %>%
#   filter(GO.ID != "unknown")
```

Remove the annotation information for genes not present in our data 
```{r}
filt_annot_GO <- subset(annot_GO, gene_id %in% filt_counts$gene_id)

str(filt_annot_GO)

# Convert filt_annot_GO from tibble to df 
filt_annot_GO <-as.data.frame(filt_annot_GO)
```


Perform GOSeq
```{r}
GO.wall<-goseq::goseq(DEG.pwf, ID_vector, gene2cat=filt_annot_GO, method="Wallenius", use_genes_without_cat=T)

GO <- GO.wall[order(GO.wall$over_represented_pvalue),]
write.csv(GO, file = "../output/functional_enrichment/GOSeq_GO_all.csv")

class(GO.wall)
head(GO.wall)
tail(GO.wall)
nrow(GO.wall)
```

Worked! 

Filter by significantly enriched GO terms (p < 0.05)
```{r}
GO_05 <- GO %>%
  dplyr::filter(over_represented_pvalue<0.05) %>%
        dplyr::arrange(., ontology, over_represented_pvalue)

write.csv(GO_05, file = "../output/functional_enrichment/GOSeq_GO_05.csv")
```

Plot all ontologies (BP, CC, MF) and order by p-value 
```{r}
GO_05_plot <- GO_05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_text(aes(label = over_represented_pvalue), hjust = -1, vjust = 0, size = 2) +
  geom_point(size=1, aes(colour = ontology)) +
  coord_flip() +
  ylim(0,305) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); GO_05_plot #Set the plot background #set title attributes

ggsave("../output/functional_enrichment/GOSeq_GO_05.pdf", GO_05_plot, width = 20, height = 10)
ggsave("../output/functional_enrichment/GOSeq_GO_05.png", GO_05_plot, width = 20, height = 10)
```

Plot ontologies (BP, CC, MF) separately and order by p-value 
```{r}
## BP 
GO_05_plot_BP <- GO_05 %>% 
  drop_na(ontology) %>% 
  filter(ontology == "BP") %>%
  mutate(term = fct_reorder(term, numDEInCat)) %>%
  #mutate(term = fct_reorder(term)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_text(aes(label = over_represented_pvalue), hjust = -1, vjust = 0, size = 2) +
  geom_point(size=1) +
  coord_flip() +
  ylim(0,305) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); GO_05_plot_BP #Set the plot background #set title attributes

ggsave("../output/functional_enrichment/GOSeq_GO_05_BP.pdf", GO_05_plot_BP, width = 30, height = 10)
ggsave("../output/functional_enrichment/GOSeq_GO_05_BP.png", GO_05_plot_BP, width = 30, height = 10)

## CC
GO_05_plot_CC <- GO_05 %>% 
  drop_na(ontology) %>% 
  filter(ontology == "CC") %>%
  mutate(term = fct_reorder(term, numDEInCat)) %>%
  #mutate(term = fct_reorder(term)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_text(aes(label = over_represented_pvalue), hjust = -1, vjust = 0, size = 2) +
  geom_point(size=1) +
  coord_flip() +
  ylim(0,305) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); GO_05_plot_CC #Set the plot background #set title attributes

ggsave("../output/functional_enrichment/GOSeq_GO_05_CC.pdf", GO_05_plot_CC, width = 30, height = 10)
ggsave("../output/functional_enrichment/GOSeq_GO_05_CC.png", GO_05_plot_CC, width = 30, height = 10)

## MF
GO_05_plot_MF <- GO_05 %>% 
  drop_na(ontology) %>% 
  filter(ontology == "MF") %>%
  mutate(term = fct_reorder(term, numDEInCat)) %>%
  #mutate(term = fct_reorder(term)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_text(aes(label = over_represented_pvalue), hjust = -1, vjust = 0, size = 2) +
  geom_point(size=1) +
  coord_flip() +
  ylim(0,305) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); GO_05_plot_MF #Set the plot background #set title attributes

ggsave("../output/functional_enrichment/GOSeq_GO_05_MF.pdf", GO_05_plot_MF, width = 30, height = 10)
ggsave("../output/functional_enrichment/GOSeq_GO_05_MF.png", GO_05_plot_MF, width = 30, height = 10)
```

Merge significant GO terms and annot list together 
```{r}
GO_genes <- annot_GO %>%
  inner_join(GO_05, by = c("GO.ID" = "category"))

length(unique(GO_genes$gene_id)) # 3786 genes with over-represented GO terms 
```

Read in DEGs and merge that data with significant GO terms 
```{r}
deg <- read.csv("../output/DEG/Unique_DEGs.csv") %>%
  rename("gene_id" = "X")

# Merge DEGs and biomin gene information
GO_deg <- GO_genes %>%
  inner_join(deg, by = "gene_id")

length(unique(GO_deg$gene_id)) # 178 DEGs with over-represented GO terms 

write.csv(GO_deg, file = "../output/functional_enrichment/DEG_GO.csv")
```

How does this relate to our DEG comparisons? Merge high v control DEGs with GO information
```{r}
## High vs control treatment comparison 
h_v_c <- read.csv("../output/DEG/High_v_Control_DEGs.csv") %>%
  rename("gene_id" = "X")

# Merge DEGs and GO information and remove extra columns
GO_h_v_c <- h_v_c %>%
  inner_join(GO_deg, by = "gene_id") %>%
  dplyr::select(-c("A.y", "B.y", "C.y", "D.y", "E.y", "F.y", "G.y", "H.y", "L.y", "M.y", "N.y"))

length(unique(GO_h_v_c$gene_id))

write.csv(GO_h_v_c, file = "../output/functional_enrichment/High_v_Control_GO.csv")
```

Merge Mid v control DEGs with GO information
```{r}
## Mid vs control treatment comparison 
m_v_c <- read.csv("../output/DEG/Mid_v_Control_DEGs.csv") %>%
  rename("gene_id" = "X")

# Merge DEGs and GO information and remove extra columns
GO_m_v_c <- m_v_c %>%
  inner_join(GO_deg, by = "gene_id") %>%
  dplyr::select(-c("A.y", "B.y", "C.y", "D.y", "E.y", "F.y", "G.y", "H.y", "L.y", "M.y", "N.y"))

length(unique(GO_m_v_c$gene_id))

write.csv(GO_m_v_c, file = "../output/functional_enrichment/Mid_v_Control_GO.csv")
```

Merge high v mid DEGs with GO information
```{r}
## High vs mid treatment comparison 
h_v_m <- read.csv("../output/DEG/High_v_Mid_DEGs.csv") %>%
  rename("gene_id" = "X")

# Merge DEGs and GO information and remove extra columns
GO_h_v_m <- h_v_m %>%
  inner_join(GO_deg, by = "gene_id") %>%
  dplyr::select(-c("A.y", "B.y", "C.y", "D.y", "E.y", "F.y", "G.y", "H.y", "L.y", "M.y", "N.y"))

length(unique(GO_h_v_m$gene_id))

write.csv(GO_h_v_m, file = "../output/functional_enrichment/High_v_Mid_GO.csv")
```

There is only 1 gene that has GO terms between Mid v Control corals (	
Pocillopora_acuta_HIv2___TS.g21597.t2). On the other hand, there are 137 and 114 genes with significant GO terms between H v C and H v M corals, respectively. 

Let's visualize these data. First, High v Control. Positive LFC means that the genes are upregulated in the high treatment. 
```{r}
GO_BP_High_v_Control <- GO_h_v_c %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "BP") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_BP_High_v_Control
ggsave("../output/functional_enrichment/GO_BP_High_v_Control_top50_pvalue.pdf", GO_BP_High_v_Control, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_BP_High_v_Control_top50_pvalue.png", GO_BP_High_v_Control, width = 30, height = 10)

GO_CC_High_v_Control <- GO_h_v_c %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "CC") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_CC_High_v_Control
ggsave("../output/functional_enrichment/GO_CC_High_v_Control_top50_pvalue.pdf", GO_CC_High_v_Control, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_CC_High_v_Control_top50_pvalue.png", GO_CC_High_v_Control, width = 30, height = 10)

GO_MF_High_v_Control <- GO_h_v_c %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "MF") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=log2FoldChange, 
               size=pvalue)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_MF_High_v_Control
ggsave("../output/functional_enrichment/GO_MF_High_v_Control_top50_pvalue.pdf", GO_MF_High_v_Control, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_MF_High_v_Control_top50_pvalue.png", GO_MF_High_v_Control, width = 30, height = 10)
```

Mid v Control. Positive LFC means that the genes are upregulated in the mid treatment. Mid v Control only shares 1 gene with GO terms and all GO terms are BP. 
```{r}
GO_BP_Mid_v_Control <- GO_m_v_c %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "BP") %>%
    #dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_BP_Mid_v_Control
ggsave("../output/functional_enrichment/GO_BP_Mid_v_Control_top50_pvalue.pdf", GO_BP_Mid_v_Control, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_BP_Mid_v_Control_top50_pvalue.png", GO_BP_Mid_v_Control, width = 30, height = 10)
```

High v Mid Positive LFC means that the genes are upregulated in the high treatment. 
```{r}
GO_BP_High_v_Mid <- GO_h_v_m %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "BP") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_BP_High_v_Mid
ggsave("../output/functional_enrichment/GO_BP_High_v_Mid_top50_pvalue.pdf", GO_BP_High_v_Mid, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_BP_High_v_Mid_top50_pvalue.png", GO_BP_High_v_Mid, width = 30, height = 10)

GO_CC_High_v_Mid <- GO_h_v_m %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "CC") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_CC_High_v_Mid
ggsave("../output/functional_enrichment/GO_CC_High_v_Mid_top50_pvalue.pdf", GO_CC_High_v_Mid, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_CC_High_v_Mid_top50_pvalue.png", GO_CC_High_v_Mid, width = 30, height = 10)

GO_MF_High_v_Mid <- GO_h_v_m %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "MF") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_MF_High_v_Mid
ggsave("../output/functional_enrichment/GO_MF_High_v_Mid_top50_pvalue.pdf", GO_MF_High_v_Mid, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_MF_High_v_Mid_top50_pvalue.png", GO_MF_High_v_Mid, width = 30, height = 10)
```

Join the GO High v Control df and GO High v Mid df by gene
```{r}
# Remove extra info
# GO_h_v_c <- GO_h_v_c %>%
#     dplyr::select(-c("baseMean", "lfcSE", "stat", "A.x", "B.x", "C.x", "D.x", "E.x", "F.x", "G.x", "H.x", "L.x", "M.x", "N.x", "under_represented_pvalue")) %>%
#     rename_with(~ paste0("h_v_c_", .), everything())
# 
# 
# GO_h_v_m <- GO_h_v_m %>%
#     dplyr::select(-c("baseMean", "lfcSE", "stat", "A.x", "B.x", "C.x", "D.x", "E.x", "F.x", "G.x", "H.x", "L.x", "M.x", "N.x", "under_represented_pvalue")) %>%
#     rename_with(~ paste0("h_v_m_", .), everything())
# 
# 
# test <- GO_h_v_c %>%
#   inner_join(GO_h_v_m, by = c("h_v_c_gene_id" = "h_v_m_gene_id"))
```

Getting this warming: Warning: Detected an unexpected many-to-many relationship between `x` and `y`.
. How can a gene have different enriched GO terms depending on the treatment?? 

Manual comparisons for GO High v Control df and GO High v Mid df
```{r}
GO_h_v_c <- GO_h_v_c %>%
  na.omit() %>%
  dplyr::arrange(log2FoldChange) %>%
  print()

GO_h_v_m <- GO_h_v_m %>%
  na.omit() %>%
  dplyr::arrange(log2FoldChange) %>%
  print()
```

Since there are so issues when joining the files, I'm going to look at the gene ids shared between these two treatment comparisons 
```{r}
my_list <- list(unique(GO_h_v_c$gene_id), unique(GO_h_v_m$gene_id))
venn(my_list)
Intersect <- venn(my_list, intersection=TRUE)
isect <- attr(Intersect, "intersection")
str(isect)
isect$`A:B`
```

There are 64 genes that are only present in H v C comparison, 41 genes that are only present in the H v M comparison and 73 genes shared between both comparisons. Filter the GO dfs so that we only have the shared genes. 

```{r}
GO_h_v_c_filt <- GO_h_v_c %>%
  dplyr::filter(gene_id %in% isect$`A:B`)

GO_h_v_m_filt <- GO_h_v_m %>%
  dplyr::filter(gene_id %in% isect$`A:B`)

# Check to see if gene_id columns are identical
identical(GO_h_v_c_filt$gene_id, GO_h_v_m_filt$gene_id) # they are not 
```

*NOTE*: the genes below are only those that are shared between h v c and h v m comparisons 

The top 5 genes that are downregulated in the high treatment compared to the control treatment are:

- Pocillopora_acuta_HIv2___RNAseq.g24121.t1 (LFC -5.663960)
- Pocillopora_acuta_HIv2___RNAseq.g13974.t1 (LFC -4.153118)
- Pocillopora_acuta_HIv2___TS.g25049.t1 (LFC -2.854903)
- Pocillopora_acuta_HIv2___RNAseq.g9799.t1 (LFC -1.640748)
- Pocillopora_acuta_HIv2___TS.g8000.t1 (LFC -1.557406)

The top 5 genes that are upregulated in the high treatment compared to the control treatment are:

- Pocillopora_acuta_HIv2___RNAseq.g22884.t1 (LFC 4.555041)
- Pocillopora_acuta_HIv2___TS.g23786.t1 (LFC 3.660517)
- Pocillopora_acuta_HIv2___TS.g26760.t1 (LFC 3.510781)
- Pocillopora_acuta_HIv2___RNAseq.3335_t (LFC 3.006764)
- Pocillopora_acuta_HIv2___RNAseq.g19477.t1 (LFC 2.853978)

The top 5 genes that are downregulated in the high treatment compared to the mid treatment are:

- Pocillopora_acuta_HIv2___RNAseq.g24121.t1 (LFC -5.548878)
- Pocillopora_acuta_HIv2___RNAseq.g13974.t1 (LFC -4.387855)
- Pocillopora_acuta_HIv2___TS.g25049.t1 (LFC -2.895585)
- Pocillopora_acuta_HIv2___TS.g8000.t1 (LFC -2.070186)
- Pocillopora_acuta_HIv2___TS.g30183.t1 (LFC -1.955101)

The top 5 genes that are upregulated in the high treatment compared to the mid treatment are:

- Pocillopora_acuta_HIv2___RNAseq.g22884.t1 (LFC 4.715193)
- Pocillopora_acuta_HIv2___RNAseq.3335_t (LFC 4.283870)
- Pocillopora_acuta_HIv2___TS.g23786.t1 (LFC 3.605581)
- Pocillopora_acuta_HIv2___RNAseq.g19477.t1 (LFC 3.378883)
- Pocillopora_acuta_HIv2___TS.g26760.t1 (LFC 2.899640)

There is some high overlap between the top 5 genes (top 5 in terms of LFC) of down and upregulation in high v control and mid. I think we should investigate these genes more explicitly, given that the largest expression changes are happening in these genes. 

Shared downregulated genes and general functions: 
- Pocillopora_acuta_HIv2___RNAseq.g24121.t1 - protein heterodimerization activity
- Pocillopora_acuta_HIv2___RNAseq.g13974.t1 - DNA binding, transcription regulation, protein heterodimerization activity
- Pocillopora_acuta_HIv2___TS.g25049.t1 - metabolic processes (amine, ethanolamine-containing compound, ammonium ion), oxidoreductase activity
- Pocillopora_acuta_HIv2___TS.g8000.t1 metabolic processes (organic acid, arachidonic acid, carboxylic acid, unsaturated fatty acid, oxoacid), biosynthetic processes (small molecule, carboxylic acid, icosanoid), oxidoreductase activity

Shared upregulated genes and general functions: 
- Pocillopora_acuta_HIv2___RNAseq.g22884.t1 - DNA binding, transcription regulation 
- Pocillopora_acuta_HIv2___TS.g23786.t1 - DNA binding
- Pocillopora_acuta_HIv2___TS.g26760.t1 - DNA topoisomerase/isomerase activity 
- Pocillopora_acuta_HIv2___RNAseq.g19477.t1 - protein serine/threonine phosphatase activity, peptidyl-serine dephosphorylation

Let's plot some of these genes! 

Read in the unique DEGs file, as that has the normalized counts, and the metadata file
```{r}
degs <- read.csv("../output/DEG/Unique_DEGs.csv")

meta <- read_csv("../Metadata/Pacuta_HI2022_Sequencing_Metadata.csv") %>%
  dplyr::select(Treatment, Species, Lifestage, TubeID) %>%
  mutate(Treatment = ifelse(Treatment == "control", "Control", Treatment)) %>%
  mutate(Treatment = ifelse(Treatment == "medium", "Mid", Treatment)) %>%
  mutate(Treatment = ifelse(Treatment == "high", "High", Treatment)) %>%
  rename(ID = TubeID) %>%
  arrange(ID)
```

Pivot the counts data, join with treatment column, and calculate mean + sd
```{r}
deg_means <- degs %>%
  pivot_longer(cols = -X, names_to = "Treatment", values_to = "Count") %>%
  left_join(meta, by = c("Treatment" = "ID")) %>%
  group_by(X, Treatment.y) %>%
  summarize(mean_count = mean(Count), sd_count = sd(Count))
```

Make a list of downregulated DEGs of interest and filter the df means by that list. Plot downregulated genes
```{r}
down <- list("Pocillopora_acuta_HIv2___RNAseq.g24121.t1", "Pocillopora_acuta_HIv2___RNAseq.g13974.t1", "Pocillopora_acuta_HIv2___TS.g25049.t1", "Pocillopora_acuta_HIv2___TS.g8000.t1")

# Select genes of interest 
down_filt <- deg_means %>% 
  filter(X %in% down) %>%
  rename("gene_id" = "X") %>%
  rename("Treatment" = "Treatment.y")  

# Arrange treatment order
down_filt$Treatment <- factor(down_filt$Treatment, levels = c("Control", "Mid", "High"))

# Plot 
shared_down <- ggplot(down_filt, aes(x = Treatment, y = mean_count, fill = Treatment)) +
  geom_bar(stat = "identity", position = position_dodge(width = 0.9)) +
  # geom_errorbar(aes(ymin = mean_count - sd_count, ymax = mean_count + sd_count),
  #               width = 0.2,  # Adjust width as needed
  #               position = position_dodge(width = 0.9)) + # Adjust dodge width as needed
  facet_wrap(~gene_id, scales = "free",) +
  labs(y = "Mean Normalized Counts", x = "") +
  scale_fill_manual(values = wes_palette("GrandBudapest1", n=3)) +
  #theme_minimal() +
  theme_bw() +
  theme(legend.position = "none") +
  theme(axis.text = element_text(size = 10, color="black"),
        axis.title=element_text(size=12, color="black", face="bold")); shared_down
ggsave("../output/functional_enrichment/Downregulated_shared_genes.pdf", shared_down)
ggsave("../output/functional_enrichment/Downregulated_shared_genes.png", shared_down)

# Plot w/ error bars
shared_down_error <- ggplot(down_filt, aes(x = Treatment, y = mean_count, fill = Treatment)) +
  geom_bar(stat = "identity", position = position_dodge(width = 0.9)) +
   geom_errorbar(aes(ymin = mean_count - sd_count, ymax = mean_count + sd_count),
                 width = 0.2,  # Adjust width as needed
                 position = position_dodge(width = 0.9)) + # Adjust dodge width as needed
  facet_wrap(~gene_id, scales = "free",) +
  labs(y = "Mean Normalized Counts", x = "") +
  scale_fill_manual(values = wes_palette("GrandBudapest1", n=3)) +
  #theme_minimal() +
  theme_bw() +
  theme(legend.position = "none") +
  theme(axis.text = element_text(size = 10, color="black"),
        axis.title=element_text(size=12, color="black", face="bold")); shared_down_error
ggsave("../output/functional_enrichment/Downregulated_shared_genes_WithErrorBars.pdf", shared_down_error)
ggsave("../output/functional_enrichment/Downregulated_shared_genes_WithErrorBars.png", shared_down_error)
```

The error bars are massive for these counts, so the plots look weird with the error bars. I have a feeling that a reviewer will want to see them though. 

Make a list of upregulated DEGs of interest and filter the df means by that list. Plot upregulated genes
```{r}
up <- list("Pocillopora_acuta_HIv2___RNAseq.g22884.t1", "Pocillopora_acuta_HIv2___TS.g23786.t1", "Pocillopora_acuta_HIv2___TS.g26760.t1", "Pocillopora_acuta_HIv2___RNAseq.g19477.t1")

# Select genes of interest 
up_filt <- deg_means %>% 
  filter(X %in% up) %>%
  rename("gene_id" = "X") %>%
  rename("Treatment" = "Treatment.y")  

# Arrange treatment order
up_filt$Treatment <- factor(up_filt$Treatment, levels = c("Control", "Mid", "High"))

# Plot 
shared_up <- ggplot(up_filt, aes(x = Treatment, y = mean_count, fill = Treatment)) +
  geom_bar(stat = "identity", position = position_dodge(width = 0.9)) +
  # geom_errorbar(aes(ymin = mean_count - sd_count, ymax = mean_count + sd_count),
  #               width = 0.2,  # Adjust width as needed
  #               position = position_dodge(width = 0.9)) + # Adjust dodge width as needed
  facet_wrap(~gene_id, scales = "free",) +
  labs(y = "Mean Normalized Counts", x = "") +
  scale_fill_manual(values = wes_palette("GrandBudapest1", n=3)) +
  #theme_minimal() +
  theme_bw() +
  theme(legend.position = "none") +
  theme(axis.text = element_text(size = 10, color="black"),
        axis.title=element_text(size=12, color="black", face="bold")); shared_up
ggsave("../output/functional_enrichment/Upregulated_shared_genes.pdf", shared_up)
ggsave("../output/functional_enrichment/Upregulated_shared_genes.png", shared_up)

# Plot w/ error bars
shared_up_error <- ggplot(up_filt, aes(x = Treatment, y = mean_count, fill = Treatment)) +
  geom_bar(stat = "identity", position = position_dodge(width = 0.9)) +
   geom_errorbar(aes(ymin = mean_count - sd_count, ymax = mean_count + sd_count),
                 width = 0.2,  # Adjust width as needed
                 position = position_dodge(width = 0.9)) + # Adjust dodge width as needed
  facet_wrap(~gene_id, scales = "free",) +
  labs(y = "Mean Normalized Counts", x = "") +
  scale_fill_manual(values = wes_palette("GrandBudapest1", n=3)) +
  #theme_minimal() +
  theme_bw() +
  theme(legend.position = "none") +
  theme(axis.text = element_text(size = 10, color="black"),
        axis.title=element_text(size=12, color="black", face="bold")); shared_up_error
ggsave("../output/functional_enrichment/Upregulated_shared_genes_WithErrorBars.pdf", shared_up_error)
ggsave("../output/functional_enrichment/Upregulated_shared_genes_WithErrorBars.png", shared_up_error)
```

Error bars for upregulated genes are not as extreme as downregulated genes. 























Optional, need to discuss w/ co-authors: 
I can also run GOSeq on the eggNOG, KEGG and PFAM terms. There are several columns for KEGG information in the `annot` file (KEGG_ko, KEGG_Pathway, KEGG_Module, KEGG_Reaction, KEGG_rclass, KEGG_TC) and I'm not sure what they all mean or which one to use in the analysis. There is also a BRITE column that looks like there is KEGG information in it. 

Select PFAM columns and split so that there is only one PFAM term per row
```{r}
annot_pfam <- annot %>%
  dplyr::select(X.query, PFAMs) %>%
  mutate_all(~if_else(. == "-", NA_character_, .)) %>%
  separate_rows(PFAMs, sep = ",") %>%
  rename(gene_id = X.query,
         PFAM.ID = PFAMs)

# Remove leading and trailing whitespaces
annot_pfam$PFAM.ID <- trimws(annot_pfam$PFAM.ID)

# Replace NAs with unknown
annot_pfam$PFAM.ID <- replace_na(annot_pfam$PFAM.ID, "unknown")

# Set gene and go ids as factors 
annot_pfam$gene_id <- as.factor(annot_pfam$gene_id)
annot_pfam$PFAM.ID <- as.factor(annot_pfam$PFAM.ID)

# Select only unique rows
annot_pfam <- unique(annot_pfam)

# Look at unique gene and go ids 
length(unique(annot_pfam$gene_id))
length(unique(annot_pfam$PFAM.ID))

# Remove rows with unknowns
# annot_GO <- annot_GO %>%
#   filter(GO.ID != "unknown")
```

Remove the annotation information for genes not present in our data 
```{r}
filt_annot_pfam <- subset(annot_pfam, gene_id %in% filt_counts$gene_id)

str(filt_annot_pfam)

# Convert filt_annot_GO from tibble to df 
filt_annot_pfam <-as.data.frame(filt_annot_pfam)
```

Perform GOSeq
```{r}
GO.wall_pfam<-goseq::goseq(DEG.pwf, ID_vector, gene2cat=filt_annot_pfam, method="Wallenius", use_genes_without_cat=T)

pfam <- GO.wall_pfam[order(GO.wall_pfam$over_represented_pvalue),]
write.csv(pfam, file = "../output/functional_enrichment/GOSeq_PFAM_all.csv")

class(GO.wall_pfam)
head(GO.wall_pfam)
tail(GO.wall_pfam)
nrow(GO.wall_pfam)
```

Filter by significantly enriched PFAM terms (p < 0.05)
```{r}
pfam_05 <- pfam %>%
  dplyr::filter(over_represented_pvalue<0.05) %>%
        dplyr::arrange(., over_represented_pvalue)

write.csv(pfam_05, file = "../output/functional_enrichment/GOSeq_PFAM_05.csv")
```

Plot and order by p-value
```{r}
pfam_05_plot <- pfam_05 %>% 
  #drop_na(ontology) %>% 
  #filter(ontology == "MF") %>%
  mutate(category = fct_reorder(category, numDEInCat)) %>%
  #mutate(term = fct_reorder(term)) %>%
  ggplot( aes(x=category, y=numDEInCat) ) +
  geom_segment( aes(x=category ,xend=category, y=0, yend=numDEInCat), color="grey") +
  geom_text(aes(label = over_represented_pvalue), hjust = -1, vjust = 0, size = 2) +
  geom_point(size=1) +
  coord_flip() +
  ylim(0,55) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); pfam_05_plot #Set the plot background #set title attributes

ggsave("../output/functional_enrichment/GOSeq_PFAM_05.pdf", pfam_05_plot, width = 30, height = 10)
ggsave("../output/functional_enrichment/GOSeq_PFAM_05.png", pfam_05_plot, width = 30, height = 10)
```

Still need to look at DEGs in the context of PFAM results and run goseq for KEGG and eggNOG annotation terms


