---
title: "Enrichment analysis"
author: "Jill Ashey"
date: "`r Sys.Date()`"
output: html_document
---

This script will perform functional enrichment analysis with GO and KEGG terms using the [identified DEGs](https://github.com/fscucchia/Hawaii2022_pH_Temp_Mcap_Pacu/blob/main/scripts/DESeq2.Rmd) from the Pacuta 2022 HI experiment. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load libraries
library("DESeq2")
library("tidyverse")
library("dplyr")
library("pheatmap")
library("RColorBrewer")
library("genefilter")
library("ggplot2")
library("gplots")
library("limma")
library("spdep") 
library("adegenet") 
library("goseq")
library("gridExtra")
library("clusterProfiler")
library("DataCombine")
library("VennDiagram")
library(rtracklayer)
library(ggridges)
#library(rrvgo)
library(GO.db)
library(tidyr)
library(forcats)
library(wesanderson)
library(GSEABase)
library(directlabels)
library(ggrepel)
library(DOSE)

#devtools::install_github("krlmlr/ulimit")
#library(ulimit)

#ulimit::memory_limit(7970592)
```

Read in all expressed genes
```{r}
filt_counts <- read.csv("../output/DEG/filtered_counts.csv")
colnames(filt_counts)[1] <- "gene_id"

str(filt_counts)
```

Read in DEGs
```{r}
degs <- read.csv("../output/DEG/Unique_DEGs.csv")
colnames(degs)[1] <- "gene_id"

str(degs)
```

Calculate read lengths from Pacuta V2 [cds file](http://cyanophora.rutgers.edu/Pocillopora_acuta/). I did this on my local computer using the code below. 
```{bash}
cd /Users/jillashey/Desktop/GFFs/pacuta/V2

awk 'BEGIN{FS="[> ]"} /^>/{val=$2;next}  {print val,length($0)}' Pocillopora_acuta_HIv2.genes.cds.fna > gene_length.txt
```

Read in gene lengths 
```{r}
length <- read.csv("../output/DEG/gene_length.txt", header = F)

length <- length %>%
  separate(V1, into = c("gene_id", "length"), sep = " ", remove = FALSE) %>%
  dplyr::select(-V1)

str(length)
```

Merge length data with filtered count data
```{r}
length_merge <- merge(length, filt_counts, by = "gene_id")

str(length_merge)
```

Read in annotation file 
```{r}
annot <- read.delim("../Metadata/Pocillopora_acuta_HIv2.genes.EggNog_results.txt", header = T) %>%
    rename(gene_id = X.query)

names(annot)
```

This functional annotation has 16,784 genes. There are 33,370 predicted protein-coding genes in the Pacuta genome, according to [Stephens et al. 2022](https://academic.oup.com/gigascience/article/doi/10.1093/gigascience/giac098/6815755). Therefore, the annotated protein coding genes represent approximately 50% of the total predicted protein coding genes described in the publication. 

Match up genes in filtered gene count list to annotation file
```{r}
probes <- filt_counts$gene_id

# Match genes in filtered count data to genes in annotation file 
probes2annot <- match(probes, annot$gene_id)

# The following is the number of probes without annotation
sum(is.na(probes2annot))
row_nas<-which(is.na(probes2annot))

# View expressed genes that do not have a match in the annotation file 
missing<-as.data.frame(probes[row_nas])
```

There are 9,887 genes that were expressed in our data that are not present in the annotation file. That is ~41% of the total genes expressed. 

Reduce annotation file to only contain genes detected in our dataset
```{r}
filtered_annot <- annot[annot$gene_id %in% probes, ]
dim(filtered_annot)
```

The annotation file now only contains genes that were detected in our dataset that have annotation information. This is 14298 genes out of the 24185 expressed genes in our dataset. 

Add in length to the annotation file 
```{r}
filtered_annot <- filtered_annot %>%
  inner_join(length, by = "gene_id")

which(is.na(filtered_annot$length)) #all genes have lengths 
```

All genes have length information. 

Format GO terms to remove 
```{r}
filtered_annot_go <- filtered_annot %>%
  dplyr::select(gene_id, GOs) %>%
  mutate_all(~if_else(. == "-", NA_character_, .)) %>%
  separate_rows(GOs, sep = ",")

# Remove leading and trailing whitespaces
filtered_annot_go$GOs <- trimws(filtered_annot_go$GOs)

sum(is.na(filtered_annot_go$GOs)) # some genes have NAs 

# Convert filtered_annot_go from tibble to df 
filtered_annot_go <-as.data.frame(filtered_annot_go)
```

GOseq requires a vector of all expressed genes, all differentially expressed genes, and gene lengths. Make vectors 
```{r}
# Generate vector with names of all expressed genes detected in our dataset 
ALL.vector <- c(filtered_annot$gene_id)

# Generate length vector with names of all expressed genes detected in our dataset 
LENGTH.vector <- as.integer(filtered_annot$length)

# Generate DEG vector 
DEG.vector <- c(degs$gene_id)

# Construct list of genes with 1 for genes in contrast and 0 for genes not in the contrast
gene.vector=as.integer(ALL.vector %in% DEG.vector) 
names(gene.vector)<-ALL.vector#set names
```

Calculate probability weighting function. Ie weight gene vector by bias for length of gene: load in the gene.vector(a list of all genes with 1 indicating it is a DEG in the group of interest and 0 meaning it is not a DEG) and bias.data (list of lengths for all genes)
```{r}
pwf <- nullp(gene.vector, bias.data = LENGTH.vector)
```

Run goseq using Wallenius method for all categories of GO terms: load in the pwf object (gene name, a 1 or 0 to indicate DEG, the length, and the pwf function; generated from pwf above), and the list of GO terms for all genes
```{r}
GO.wall<-goseq(pwf, gene2cat=filtered_annot_go, method="Wallenius", use_genes_without_cat=TRUE)

GO <- GO.wall[order(GO.wall$over_represented_pvalue),]
colnames(GO)[1] <- "GOterm"

write.csv(GO, file = "../output/functional_enrichment/GOSeq_GO_all.csv")
```

Adjust pvalues for multiple hypotheses testing 
```{r}
GO$bh_adjust <- p.adjust(GO$over_represented_pvalue, method = "BH")
```

When applying multiple hypotheses testing, all GO terms are not significant. I am going to move forward with the filtering using the non-adjusted pvalue, as I do not think that MHT is necessarily needed here if we are not doing a 1:1 test. 

Filter by significantly enriched GO terms (p < 0.05)
```{r}
GO_05 <- GO %>%
  dplyr::filter(over_represented_pvalue<0.05) %>%
        dplyr::arrange(., ontology, over_represented_pvalue)

write.csv(GO_05, file = "../output/functional_enrichment/GOSeq_GO_05.csv")
```

Plot all ontologies (BP, CC, MF) and order by p-value 
```{r}
GO_05_plot <- GO_05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_text(aes(label = over_represented_pvalue), hjust = -1, vjust = 0, size = 2) +
  geom_point(size=1, aes(colour = ontology)) +
  coord_flip() +
  ylim(0,305) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); GO_05_plot #Set the plot background #set title attributes

ggsave("../output/functional_enrichment/GOSeq_GO_05.pdf", GO_05_plot, width = 20, height = 10)
ggsave("../output/functional_enrichment/GOSeq_GO_05.png", GO_05_plot, width = 20, height = 10)
```

Plot ontologies (BP, CC, MF) separately and order by p-value 
```{r}
## BP 
GO_05_plot_BP <- GO_05 %>% 
  drop_na(ontology) %>% 
  filter(ontology == "BP") %>%
  mutate(term = fct_reorder(term, numDEInCat)) %>%
  #mutate(term = fct_reorder(term)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_text(aes(label = over_represented_pvalue), hjust = -1, vjust = 0, size = 2) +
  geom_point(size=1) +
  coord_flip() +
  ylim(0,305) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); GO_05_plot_BP #Set the plot background #set title attributes

ggsave("../output/functional_enrichment/GOSeq_GO_05_BP.pdf", GO_05_plot_BP, width = 30, height = 10)
ggsave("../output/functional_enrichment/GOSeq_GO_05_BP.png", GO_05_plot_BP, width = 30, height = 10)

## CC
GO_05_plot_CC <- GO_05 %>% 
  drop_na(ontology) %>% 
  filter(ontology == "CC") %>%
  mutate(term = fct_reorder(term, numDEInCat)) %>%
  #mutate(term = fct_reorder(term)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_text(aes(label = over_represented_pvalue), hjust = -1, vjust = 0, size = 2) +
  geom_point(size=1) +
  coord_flip() +
  ylim(0,305) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); GO_05_plot_CC #Set the plot background #set title attributes

ggsave("../output/functional_enrichment/GOSeq_GO_05_CC.pdf", GO_05_plot_CC, width = 30, height = 10)
ggsave("../output/functional_enrichment/GOSeq_GO_05_CC.png", GO_05_plot_CC, width = 30, height = 10)

## MF
GO_05_plot_MF <- GO_05 %>% 
  drop_na(ontology) %>% 
  filter(ontology == "MF") %>%
  mutate(term = fct_reorder(term, numDEInCat)) %>%
  #mutate(term = fct_reorder(term)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_text(aes(label = over_represented_pvalue), hjust = -1, vjust = 0, size = 2) +
  geom_point(size=1) +
  coord_flip() +
  ylim(0,305) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); GO_05_plot_MF #Set the plot background #set title attributes

ggsave("../output/functional_enrichment/GOSeq_GO_05_MF.pdf", GO_05_plot_MF, width = 30, height = 10)
ggsave("../output/functional_enrichment/GOSeq_GO_05_MF.png", GO_05_plot_MF, width = 30, height = 10)
```

Merge significant GO terms and annot list together 
```{r}
GO_genes <- filtered_annot_go %>%
  inner_join(GO_05, by = c("GOs" = "GOterm"))

length(unique(GO_genes$gene_id)) # 5931 genes with over-represented GO terms 
```

Read in DEGs and merge that data with significant GO terms 
```{r}
deg <- read.csv("../output/DEG/Unique_DEGs.csv") %>%
  rename("gene_id" = "X")

# Merge DEGs and biomin gene information
GO_deg <- GO_genes %>%
  inner_join(deg, by = "gene_id")

length(unique(GO_deg$gene_id)) # 204 DEGs with over-represented GO terms 

write.csv(GO_deg, file = "../output/functional_enrichment/DEG_GO.csv")
```

How does this relate to our DEG comparisons? Merge high v control DEGs with GO information
```{r}
## High vs control treatment comparison 
h_v_c <- read.csv("../output/DEG/High_v_Control_DEGs.csv") %>%
  rename("gene_id" = "X")

# Merge DEGs and GO information and remove extra columns
GO_h_v_c <- h_v_c %>%
  inner_join(GO_deg, by = "gene_id") %>%
  dplyr::select(-c("A.y", "B.y", "C.y", "D.y", "E.y", "F.y", "G.y", "H.y", "L.y", "M.y", "N.y"))

length(unique(GO_h_v_c$gene_id))

write.csv(GO_h_v_c, file = "../output/functional_enrichment/High_v_Control_GO.csv")
```

Merge Mid v control DEGs with GO information
```{r}
## Mid vs control treatment comparison 
m_v_c <- read.csv("../output/DEG/Mid_v_Control_DEGs.csv") %>%
  rename("gene_id" = "X")

# Merge DEGs and GO information and remove extra columns
GO_m_v_c <- m_v_c %>%
  inner_join(GO_deg, by = "gene_id") %>%
  dplyr::select(-c("A.y", "B.y", "C.y", "D.y", "E.y", "F.y", "G.y", "H.y", "L.y", "M.y", "N.y"))

length(unique(GO_m_v_c$gene_id))

write.csv(GO_m_v_c, file = "../output/functional_enrichment/Mid_v_Control_GO.csv")
```

Merge high v mid DEGs with GO information
```{r}
## High vs mid treatment comparison 
h_v_m <- read.csv("../output/DEG/High_v_Mid_DEGs.csv") %>%
  rename("gene_id" = "X")

# Merge DEGs and GO information and remove extra columns
GO_h_v_m <- h_v_m %>%
  inner_join(GO_deg, by = "gene_id") %>%
  dplyr::select(-c("A.y", "B.y", "C.y", "D.y", "E.y", "F.y", "G.y", "H.y", "L.y", "M.y", "N.y"))

length(unique(GO_h_v_m$gene_id))

write.csv(GO_h_v_m, file = "../output/functional_enrichment/High_v_Mid_GO.csv")
```

Let's visualize these data. First, High v Control. Positive LFC means that the genes are upregulated in the high treatment. 
```{r}
GO_BP_High_v_Control <- GO_h_v_c %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "BP") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_BP_High_v_Control
ggsave("../output/functional_enrichment/GO_BP_High_v_Control_top50_pvalue.pdf", GO_BP_High_v_Control, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_BP_High_v_Control_top50_pvalue.png", GO_BP_High_v_Control, width = 30, height = 10)

GO_CC_High_v_Control <- GO_h_v_c %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "CC") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_CC_High_v_Control
ggsave("../output/functional_enrichment/GO_CC_High_v_Control_top50_pvalue.pdf", GO_CC_High_v_Control, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_CC_High_v_Control_top50_pvalue.png", GO_CC_High_v_Control, width = 30, height = 10)

GO_MF_High_v_Control <- GO_h_v_c %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "MF") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=log2FoldChange, 
               size=pvalue)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_MF_High_v_Control
ggsave("../output/functional_enrichment/GO_MF_High_v_Control_top50_pvalue.pdf", GO_MF_High_v_Control, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_MF_High_v_Control_top50_pvalue.png", GO_MF_High_v_Control, width = 30, height = 10)
```

Mid v Control. Positive LFC means that the genes are upregulated in the mid treatment. Mid v Control only shares 1 gene with GO terms and all GO terms are BP. 
```{r}
GO_BP_Mid_v_Control <- GO_m_v_c %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "BP") %>%
    #dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_BP_Mid_v_Control
ggsave("../output/functional_enrichment/GO_BP_Mid_v_Control_top50_pvalue.pdf", GO_BP_Mid_v_Control, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_BP_Mid_v_Control_top50_pvalue.png", GO_BP_Mid_v_Control, width = 30, height = 10)
```

High v Mid Positive LFC means that the genes are upregulated in the high treatment. 
```{r}
GO_BP_High_v_Mid <- GO_h_v_m %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "BP") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_BP_High_v_Mid
ggsave("../output/functional_enrichment/GO_BP_High_v_Mid_top50_pvalue.pdf", GO_BP_High_v_Mid, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_BP_High_v_Mid_top50_pvalue.png", GO_BP_High_v_Mid, width = 30, height = 10)

GO_CC_High_v_Mid <- GO_h_v_m %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "CC") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_CC_High_v_Mid
ggsave("../output/functional_enrichment/GO_CC_High_v_Mid_top50_pvalue.pdf", GO_CC_High_v_Mid, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_CC_High_v_Mid_top50_pvalue.png", GO_CC_High_v_Mid, width = 30, height = 10)

GO_MF_High_v_Mid <- GO_h_v_m %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "MF") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=pvalue, 
               size=log2FoldChange)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="p value", size="L2FC"); GO_MF_High_v_Mid
ggsave("../output/functional_enrichment/GO_MF_High_v_Mid_top50_pvalue.pdf", GO_MF_High_v_Mid, width = 30, height = 10)
ggsave("../output/functional_enrichment/GO_MF_High_v_Mid_top50_pvalue.png", GO_MF_High_v_Mid, width = 30, height = 10)
```

Join the GO High v Control df and GO High v Mid df by gene
```{r}
# Remove extra info
# GO_h_v_c <- GO_h_v_c %>%
#     dplyr::select(-c("baseMean", "lfcSE", "stat", "A.x", "B.x", "C.x", "D.x", "E.x", "F.x", "G.x", "H.x", "L.x", "M.x", "N.x", "under_represented_pvalue")) %>%
#     rename_with(~ paste0("h_v_c_", .), everything())
# 
# 
# GO_h_v_m <- GO_h_v_m %>%
#     dplyr::select(-c("baseMean", "lfcSE", "stat", "A.x", "B.x", "C.x", "D.x", "E.x", "F.x", "G.x", "H.x", "L.x", "M.x", "N.x", "under_represented_pvalue")) %>%
#     rename_with(~ paste0("h_v_m_", .), everything())
# 
# 
# test <- GO_h_v_c %>%
#   inner_join(GO_h_v_m, by = c("h_v_c_gene_id" = "h_v_m_gene_id"))
```

Getting this warming: Warning: Detected an unexpected many-to-many relationship between `x` and `y`.
. How can a gene have different enriched GO terms depending on the treatment?? 

Manual comparisons for GO High v Control df and GO High v Mid df
```{r}
GO_h_v_c <- GO_h_v_c %>%
  na.omit() %>%
  dplyr::arrange(log2FoldChange) %>%
  print()

GO_h_v_m <- GO_h_v_m %>%
  na.omit() %>%
  dplyr::arrange(log2FoldChange) %>%
  print()
```

Since there are so issues when joining the files, I'm going to look at the gene ids shared between these two treatment comparisons 
```{r}
my_list <- list(unique(GO_h_v_c$gene_id), unique(GO_h_v_m$gene_id))
venn(my_list)
Intersect <- venn(my_list, intersection=TRUE)
isect <- attr(Intersect, "intersection")
str(isect)
isect$`A:B`
```

Filter the GO dfs so that we only have the shared genes. 
```{r}
# High v control
GO_h_v_c_filt <- GO_h_v_c %>%
  dplyr::filter(gene_id %in% isect$`A:B`)
GO_h_v_c_filt <- GO_h_v_c_filt[order(GO_h_v_c_filt$log2FoldChange),] # order from least to greatest 
unique(GO_h_v_c_filt$gene_id)

# High v mid 
GO_h_v_m_filt <- GO_h_v_m %>%
  dplyr::filter(gene_id %in% isect$`A:B`)
GO_h_v_m_filt <- GO_h_v_m_filt[order(GO_h_v_m_filt$log2FoldChange),] # order from least to greatest 
unique(GO_h_v_m_filt$gene_id)

# Check to see if gene_id columns are identical
identical(GO_h_v_c_filt$gene_id, GO_h_v_m_filt$gene_id) # they are not 
```

*NOTE*: the genes below are only those that are shared between h v c and h v m comparisons 

Plot GO terms of shared genes 
```{r}
blah <- GO_h_v_c_filt %>% 
    #mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    dplyr::filter(ontology == "MF") %>%
    dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    #top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               colour=log2FoldChange, 
               size=over_represented_pvalue)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", colour="LFC", size="overrepresented pvalue"); blah
```

The top 5 genes that are most DOWNregulated in the high treatment compared to the control treatment are:

- Pocillopora_acuta_HIv2___RNAseq.g8481.t1 (LFC -8.8221765)
- Pocillopora_acuta_HIv2___RNAseq.g12512.t1 (LFC -8.446681)
- Pocillopora_acuta_HIv2___RNAseq.22237_t (LFC -7.667258)
- Pocillopora_acuta_HIv2___RNAseq.g11960.t1 (LFC -6.917192)
- Pocillopora_acuta_HIv2___RNAseq.g24121.t1 (LFC -5.6686604)

The top 5 genes that are most DOWNregulated in the high treatment compared to the mid treatment are:

- Pocillopora_acuta_HIv2___RNAseq.g8481.t1 (LFC -8.396540)
- Pocillopora_acuta_HIv2___RNAseq.g12512.t1 (LFC -7.753341)
- Pocillopora_acuta_HIv2___RNAseq.22237_t (LFC -7.162973)
- Pocillopora_acuta_HIv2___RNAseq.g11960.t1 (LFC -6.860989)
- Pocillopora_acuta_HIv2___RNAseq.g24121.t1 (LFC -5.554880)

The top 5 genes that are most UPregulated in the high treatment compared to the control treatment are:

- Pocillopora_acuta_HIv2___RNAseq.g22884.t1 (LFC 4.549559)
- Pocillopora_acuta_HIv2___TS.g23786.t1 (LFC 3.653126)
- Pocillopora_acuta_HIv2___TS.g26760.t1 (LFC 3.505418)
- Pocillopora_acuta_HIv2___RNAseq.3335_t (LFC 3.001841)
- Pocillopora_acuta_HIv2___RNAseq.g19477.t1 (LFC 2.848650)

The top 5 genes that are most UPregulated in the high treatment compared to the mid treatment are:

- Pocillopora_acuta_HIv2___RNAseq.g22884.t1 (LFC 4.709843)
- Pocillopora_acuta_HIv2___RNAseq.3335_t (LFC 4.278247)
- Pocillopora_acuta_HIv2___TS.g23786.t1 (LFC 3.599031)
- Pocillopora_acuta_HIv2___RNAseq.g19477.t1 (LFC 3.373083)
- Pocillopora_acuta_HIv2___TS.g26760.t1 (LFC 2.893989)

Compared to mid and control corals, high corals strongly downregulated `Pocillopora_acuta_HIv2___RNAseq.g8481.t1`, `Pocillopora_acuta_HIv2___RNAseq.g12512.t1`, `Pocillopora_acuta_HIv2___RNAseq.22237_t`, `Pocillopora_acuta_HIv2___RNAseq.g11960.t1`, and `Pocillopora_acuta_HIv2___RNAseq.g24121.t1`.

Compared to mid and control corals, high corals strongly upregulated `Pocillopora_acuta_HIv2___RNAseq.g22884.t1`, `Pocillopora_acuta_HIv2___TS.g23786.t1`, `Pocillopora_acuta_HIv2___RNAseq.g19477.t1`, `Pocillopora_acuta_HIv2___RNAseq.3335_t`, and `Pocillopora_acuta_HIv2___TS.g26760.t1`. 

Given these genes have the greatest LFC in the comparisons, I think we should investigate them further. 

Plot the DEGs of interest with the GO information. First, read in DEG and GO information.
```{r}
deg_go <- read.csv("../output/functional_enrichment/DEG_GO.csv")
```

Filter by downregulated genes of interest 
```{r}
down <- c("Pocillopora_acuta_HIv2___RNAseq.g8481.t1", "Pocillopora_acuta_HIv2___RNAseq.g12512.t1", "Pocillopora_acuta_HIv2___RNAseq.22237_t", "Pocillopora_acuta_HIv2___RNAseq.g11960.t1", "Pocillopora_acuta_HIv2___RNAseq.g24121.t1")

# Select genes of interest 
down_filt <- deg_go %>% 
  filter(gene_id %in% down) %>%
  na.omit()
```

Plot GO term information for downregulated genes of interest 
```{r}
down_deg_go <- down_filt %>% 
    mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    na.omit() %>%
    #dplyr::filter(ontology == "BP") %>%
    #dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    #top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               #colour=pvalue, 
               color=over_represented_pvalue)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", size="pvalue"); down_deg_go
```

Filter by upregulated genes of interest 
```{r}
up <- c("Pocillopora_acuta_HIv2___RNAseq.g22884.t1", "Pocillopora_acuta_HIv2___TS.g23786.t1", "Pocillopora_acuta_HIv2___RNAseq.g19477.t1", "Pocillopora_acuta_HIv2___RNAseq.3335_t", "Pocillopora_acuta_HIv2___TS.g26760.t1")

# Select genes of interest 
up_filt <- deg_go %>% 
  filter(gene_id %in% up) %>%
  na.omit()
```

Plot GO term information for downregulated genes of interest 
```{r}
up_deg_go <- up_filt %>% 
    mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    na.omit() %>%
    #dplyr::filter(ontology == "BP") %>%
    #dplyr::  filter(log2FoldChange > 2 | log2FoldChange < -2) %>%
    #top_n(50, wt=over_represented_pvalue) %>% 
    ggplot(aes(x=numDEInCat, 
               y=term, 
               #colour=pvalue, 
               color=over_represented_pvalue)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Number of DEGs", y="GO term", size="pvalue"); up_deg_go
```

Add slim terms to dataset 
```{r}
GOslim <- read_csv("~/Desktop/PutnamLab/Repositories/SedimentStress/SedimentStress/Output/GOSeq/GO-GOslim.csv")
```

Merge GO slim info and log p-value for downreg genes
```{r}
down_filt_slim <- down_filt %>%
  inner_join(GOslim, by = c("GOs" = "GO_ID"))

# Log the overrepresented pvalue 
down_filt_slim <- down_filt_slim %>%
  mutate(log_over_represented_pvalue = -log10(over_represented_pvalue))

# Sort the data by GOSlim_bin and log_over_represented_pvalue
down_filt_slim <- down_filt_slim %>%
  arrange(GOSlim_bin, log_over_represented_pvalue)

# Save csv with slim info for these genes 
write.csv(down_filt_slim, file = "../output/functional_enrichment/Down_filt_genes_GOslim.csv")
```

Plot downregulated GO slim information. 
```{r}
# Set the factor levels for term.x based on the sorted order
down_filt_slim$term.x <- factor(down_filt_slim$term.x, levels = unique(down_filt_slim$term.x))

# Set the order of the facet grid levels
down_filt_slim$ontology <- factor(down_filt_slim$ontology, levels = c("BP", "MF", "CC"))

# Create a new factor for GOSlim_bin ordered by ontology
down_filt_slim <- down_filt_slim %>%
  mutate(GOSlim_bin = factor(GOSlim_bin, levels = unique(GOSlim_bin[order(ontology)])))

# Plot
down_filt_slim_plot <- down_filt_slim %>%
      ggplot(aes(x=log_over_represented_pvalue, 
               y=term.x,
               colour=ontology)) +
        geom_point(size = 20) +
    scale_color_manual(values = c("cornflowerblue", "mediumvioletred", "seagreen")) +
        #expand_limits(x=0) +
        #labs(x="Number of DEGs", y="GO term", color="pvalue") %>%
        facet_grid(GOSlim_bin ~ ., scales = "free", space='free', labeller = label_wrap_gen(width = 55, multi_line = TRUE)) +
      #xlab("Number of DEGs") +
      #ylab("") +
      labs(x="-log10(over-represented p-value)", y="") +
      theme_bw() + 
   theme(
    strip.text.y = element_text(size = 28, colour = "black", angle = 0, face = "bold"),
    axis.text = element_text(size = 32, colour = "black"),
    axis.title.x = element_text(size = 35, colour = "black"),
    legend.text = element_text(size = 24),  # Increase legend text size
    legend.title = element_text(size = 26), # Increase legend title size
    legend.key.size = unit(1.5, "lines"),   # Increase size of legend key
    panel.border = element_rect(colour = "black", size = 1.5),  # Darken and thicken the panel border
    strip.background = element_rect(colour = "black", size = 1.5) # Darken and thicken the strip background border
  )

ggsave("../output/functional_enrichment/Downregulated_shared_genes_GOSlim.pdf", down_filt_slim_plot, width = 35, height = 25)
ggsave("../output/functional_enrichment/Downregulated_shared_genes_GOSlim.png", down_filt_slim_plot, width = 35, height = 25)
```

Merge GO slim info and log p-value for upreg genes 
```{r}
up_filt_slim <- up_filt %>%
  inner_join(GOslim, by = c("GOs" = "GO_ID"))

# Log the overrepresented pvalue 
up_filt_slim <- up_filt_slim %>%
  mutate(log_over_represented_pvalue = -log10(over_represented_pvalue))

# Sort the data by GOSlim_bin and log_over_represented_pvalue
up_filt_slim <- up_filt_slim %>%
  arrange(GOSlim_bin, log_over_represented_pvalue)

# Save csv with slim info for these genes 
write.csv(up_filt_slim, file = "../output/functional_enrichment/Up_filt_genes_GOslim.csv")
```

Plot
```{r}
# Set the factor levels for term.x based on the sorted order
up_filt_slim$term.x <- factor(up_filt_slim$term.x, levels = unique(up_filt_slim$term.x))

# Set the order of the facet grid levels
up_filt_slim$ontology <- factor(up_filt_slim$ontology, levels = c("BP", "MF", "CC"))

# Create a new factor for GOSlim_bin ordered by ontology
up_filt_slim <- up_filt_slim %>%
  mutate(GOSlim_bin = factor(GOSlim_bin, levels = unique(GOSlim_bin[order(ontology)])))

# Plot
up_filt_slim_plot <- up_filt_slim %>%
      ggplot(aes(x=log_over_represented_pvalue, 
               y=term.x,
               colour=ontology)) +
        geom_point(size = 20) +
      scale_color_manual(values = c("cornflowerblue", "mediumvioletred", "seagreen")) +
        #expand_limits(x=0) +
        facet_grid(GOSlim_bin ~ ., scales = "free", space='free', labeller = label_wrap_gen(width = 55, multi_line = TRUE)) +
      #xlab("Number of DEGs") +
      #ylab("") +
      labs(x="-log10(over-represented p-value)", y="") +
      theme_bw() + 
   theme(
    strip.text.y = element_text(size = 28, colour = "black", angle = 0, face = "bold"),
    axis.text = element_text(size = 32, colour = "black"),
    axis.title.x = element_text(size = 35, colour = "black"),
    legend.text = element_text(size = 24),  # Increase legend text size
    legend.title = element_text(size = 26), # Increase legend title size
    legend.key.size = unit(1.5, "lines"),   # Increase size of legend key
    panel.border = element_rect(colour = "black", size = 1.5),  # Darken and thicken the panel border
    strip.background = element_rect(colour = "black", size = 1.5) # Darken and thicken the strip background border
  )

ggsave("../output/functional_enrichment/Upregulated_shared_genes_GOSlim.pdf", up_filt_slim_plot, width = 35, height = 25)
ggsave("../output/functional_enrichment/Upregulated_shared_genes_GOSlim.png", up_filt_slim_plot, width = 35, height = 25)
```














Trying out ridgeplot
```{r}
data("geneList")
x <- gseDO(geneList)
ridgeplot(x) + labs(x = "enrichment distribution")

```



